---
title: "MapMet project: PT analysis from Patel et al., 2024"
author: "Sara Wernig-Zorc"
date: "`r Sys.Date()`"
output:
  pdf_document:
    toc: yes
    toc_depth: '5'
  html_document:
    toc: yes
    toc_depth: 5
    toc_float: yes
    collapsed: no
    code_folding: hide
    highlight: pygments
    theme: spacelab
    df_print: paged
params:
  input: /home/rstudio/mnt_data/
  out: /home/rstudio/mnt_out/patel2024/
  data: /home/rstudio/mnt_data/
  nb_tumors: /home/rstudio/mnt_nb_tumors/
  GO_db: /home/rstudio/mnt_resources/GSEA_GeneSetEnrichments/
  markers: /home/rstudio/mnt_resources/MES_ADRN/
---

```{r setup, include=FALSE, echo = FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      tidy = TRUE, 
                      error = FALSE, 
                      messages = FALSE,
                      warning = FALSE)

library(parallel)
future::plan(strategy = 'future::multicore', workers = 20)
options(future.globals.maxSize = 99 * 10^9)

print(params)
```

```{r R packages: load, include=FALSE, echo=FALSE}
library(Seurat)
library(scCustomize)
library(dplyr)
library(ggplot2)
```


--------------------- 13.05.2024 ------------------

```{r input}
tumorClust <- readRDS(file = paste0(params$nb_tumors,"/patel2024.scRNA/data/scRNA-seq/patel2024_NB_cluster.Rds"))
whole_dataset <- readRDS(file = paste0(params$nb_tumors,"/patel2024.scRNA/data/scRNA-seq/20220117_final_HTAPP_neuroblastoma_LIGER.Rds"))

pdf(file = paste0(params$out,"/patel2024/UMAP_tumorCells.pdf"), height = 7,width = 7)
DimPlot(tumorClust)
dev.off()


tumorClust$original_celltypes <- Idents(tumorClust) 


```

```{r}

COLOR_CODE_IMC <- c("fibroblast/endothel"=rgb(212,89,67, maxColorValue = 255),
                  "schwann cell"=rgb(156,76,24, maxColorValue = 255),
                  "MSC"=rgb(174,129,37, maxColorValue = 255),
                  "HPC"=rgb(212,143,75, maxColorValue = 255),
                  "HSPC"=rgb(253,180,98, maxColorValue = 255),
                  "Ki67+ CXCR4+ cell"=rgb(254,215,0, maxColorValue = 255),
                  "LIN- Ki67+"=rgb(254,217,166, maxColorValue = 255),
                  "B cell"=rgb(168,128,250, maxColorValue = 255),
                  "GZMB+ S100B- DC/NK"=rgb(28,146,255, maxColorValue = 255),
                  "GZMB+ S100B+ DC/NK"=rgb(0,255,255, maxColorValue = 255),
                  "CD8+ naive T cell"=rgb(188,128,189, maxColorValue = 255),
                  "DN GZMB+ T cell"=rgb(190,174,212, maxColorValue = 255),
                  "CD8+ S100B+ T cell"=rgb(253,218,236, maxColorValue = 255),
                  "CD8+ GZMB+ PD1lo T cell"=rgb(251,154,153, maxColorValue = 255),
                  "CD4+ naive T cell"=rgb(242,3,137, maxColorValue = 255),
                  "DN T cell"=rgb(251,128,114, maxColorValue = 255),
                  "CD4+ PD1+ T cell"=rgb(254,142,175, maxColorValue = 255),
                  "CD8+ PDL1lo T cell"=rgb(240,0,240, maxColorValue = 255),
                  "CD4+ S100B+ T cell"=rgb(227,26,28, maxColorValue = 255),
                  "dense T cell region"=rgb(145,2,144, maxColorValue = 255),
                  "CD14+ PDL1- MO"=rgb(104,180,238, maxColorValue = 255),
                  "CD14+ PDL1+ MO"=rgb(51,180,180, maxColorValue = 255),
                  "CD14- PDL1- MO/DC"=rgb(39,130,187, maxColorValue = 255),
                  "CD14- PDL1+ MO/DC"=rgb(166,206,227, maxColorValue = 255),
                  "monoblast-like"=rgb(128,177,211, maxColorValue = 255),
                  "myelocyte"=rgb(217,217,217, maxColorValue = 255),          
                  "myeloblast"=rgb(2,0,138, maxColorValue = 255),
                  "neutrophil"=rgb(113,113,113, maxColorValue = 255),
                  "pDC"=rgb(212,241,240, maxColorValue = 255),
                  "band cell"=rgb(169,189,210, maxColorValue = 255),
                  "CD24+ marker-lo TC"=rgb(230,245,201, maxColorValue = 255),
                  "early SYM-like TC"=rgb(204,235,197, maxColorValue = 255),
                  "bridge-like TC"=rgb(198,255,72, maxColorValue = 255),
                  "GATA3hi TC"=rgb(65,208,57, maxColorValue = 255),
                  "CD24- marker-lo TC"=rgb(60,119,97, maxColorValue = 255),
                  "CHGAhi TC"=rgb(102,166,30, maxColorValue = 255),
                  "Ki67hi TC"=rgb(149,144,89, maxColorValue = 255),
                  "CXCR4hi TC"=rgb(210,205,126, maxColorValue = 255),
                  "CD44+ mes-like TC"=rgb(27,158,119, maxColorValue = 255),
                  "GD2lo TC"=rgb(102,194,165, maxColorValue = 255),
                  "neural progenitor"=rgb(238,118,0, maxColorValue = 255),
                  "other"=rgb(255,255,209, maxColorValue = 255))

```


```{r}
gene.list.RNA <- c("CD14","CD24","CD34","CD3E","CD4","CD44","CD74","CD8A","CHGA",
                   "CXCR4","ELAVL4","FOXP3","FUT4","GATA3","GZMB",   
                   "HLA-A","HLA-DRA","ITGAM","ITGAX","LUM","MKI67","MME", 
                   "MPO","MS4A1","NCAM1","PDCD1","PRPH","PTPRC",  
                   "S100B","SOX10","VIM","CD274","ST8SIA1","B4GALNT1")
```

```{r input data Lazic et al, eval=FALSE, include=FALSE}
#protein2gene <- read.csv("/home/mnt_data/MapMet/lazic.imc.2024/gene_list/protein2gene_2024-05-03.csv", 
#sep=";", header = TRUE)

#IMC_markers <- protein2gene %>% filter(used_for_clustering_RNA == "yes")
#gene.list <- IMC_markers$gene_name

#imc_data <- readRDS("/home/data/MapMet/lazic.imc.2024/Rds/imc_seurat.rds")
imc_data <- readRDS("~/mnt_data/tmp/imc_seurat.rds")

gene.list <- c("CD14","CD24","CD34","CD3E","CD4","CD44","CD74","CD8A","CHGA",
                   "CXCR4","ELAVL4","FOXP3","FUT4","GATA3","GZMB",   
                   "HLA-A","HLA-DRA","ITGAM","ITGAX","LUM","MKI67","MME", 
                   "MPO","MS4A1","NCAM1","PDCD1","PRPH","PTPRC",  
                   "S100B","SOX10","VIM")

gene.list 
```

```{r}

pdf(file = paste0(params$out,"/patel2024/DotPlot_IMCmarkers_tumorClust_v1.pdf"), height = 10,width = 8)
DotPlot(tumorClust, features = gene.list.RNA, 
        cols = c("gray","darkviolet")) + coord_flip()
dev.off()

pdf(file = paste0(params$out,"/patel2024/DotPlot_IMCmarkers_tumorAndImmuneClust_v1.pdf"), height = 10,width = 12)
DotPlot(whole_dataset_downsampled, features = gene.list.RNA, 
        cols = c("gray","darkviolet")) + coord_flip()
dev.off()


pdf(file = paste0(params$out,"/patel2024/DotPlot_IMCmarkers_tumorClust_v2.pdf"))
Clustered_DotPlot(tumorClust, features = gene.list.RNA)
dev.off()

pdf(file = paste0(params$out,"/patel2024/DotPlot_IMCmarkers_tumorAndImmuneClust_v2.pdf"))
Clustered_DotPlot(whole_dataset_downsampled, features = gene.list.RNA)
dev.off()
```

```{r}
# Downsample the number of cells per identity class
whole_dataset_downsampled <- subset(x = whole_dataset, downsample = 1000)
saveRDS(whole_dataset_downsampled, file = paste0(params$nb_tumors,"/patel2024.scRNA/data/scRNA-seq/patel2024_whole_dataset_downsampled_downsampled.Rds"))

tumorClust_downsampled <- subset(x = tumorClust, downsample = 1000)
saveRDS(tumorClust_downsampled, file = paste0(params$nb_tumors,"/patel2024.scRNA/data/scRNA-seq/patel2024_tumor_cluster_downsampled.Rds"))
```

```{r}
whole_dataset_downsampled <- readRDS(file = paste0(params$nb_tumors,"/patel2024.scRNA/data/scRNA-seq/patel2024_whole_dataset_downsampled.Rds"))

tumorClust_downsampled <- readRDS(file = paste0(params$nb_tumors,"/patel2024.scRNA/data/scRNA-seq/patel2024_tumor_cluster_downsampled.Rds"))
```

Clustering with IMC markers on the whole dataset (downsampled to 1000 cell per cluster)

```{r whole_dataset}
# principal component analysis
whole_dataset_downsampled <- RunPCA(whole_dataset_downsampled, 
                        assay = "RNA", 
                        npcs = 50,
                        features = gene.list.RNA)

# PCA plot
PCAPlot(object = whole_dataset_downsampled)

# find closest cells based on gene expression of the Yang2019 genes
whole_dataset_downsampled <- FindNeighbors(whole_dataset_downsampled, 
                           features = gene.list.RNA,
                           reduction = "pca", 
                           dims = 1:10,
                           k.param = 500)

# find clusters based on nearest neighbours
whole_dataset_downsampled <- FindClusters(whole_dataset_downsampled, 
                          resolution = 0.2,
                          random.seed = 28,
                          group.singletons = FALSE)

# Dimension reductions with Uniform Manifold Approximation and Projection (UMAP)
whole_dataset_downsampled <- RunUMAP(whole_dataset_downsampled, 
                     reduction = "pca", 
                     dims = 1:10,
                     n.epochs = 500,
                     min.dist = 0.1,
                     seed.use = 28)

# plot UMAP
DimPlot(whole_dataset_downsampled,
        reduction = "umap")

DoHeatmap(whole_dataset_downsampled, features = gene.list.RNA)
DotPlot(whole_dataset_downsampled, features = gene.list.RNA)
```

Clustering with IMC markers on the tumor cells only

```{r tumor_cluster_downsampled}
# normalize and scale count data
tumorClust_downsampled <- tumorClust_downsampled %>%
    NormalizeData() %>%
    ScaleData() 

# principal component analysis
tumorClust_downsampled <- RunPCA(tumorClust_downsampled, 
                        assay = "RNA", 
                        npcs = 50,
                        features = gene.list.RNA)

# PCA plot
PCAPlot(object = tumorClust_downsampled)

# find closest cells based on gene expression of the Yang2019 genes
tumorClust_downsampled <- FindNeighbors(tumorClust_downsampled, 
                           features = gene.list.RNA,
                           reduction = "pca", 
                           dims = 1:10,
                           k.param = 100) #tested 200,500

# find clusters based on nearest neighbours
tumorClust_downsampled <- FindClusters(tumorClust_downsampled, 
                          resolution = 0.8, #tested 0.2, 0.4, 0.8
                          random.seed = 28,
                          group.singletons = FALSE)

# Dimension reductions with Uniform Manifold Approximation and Projection (UMAP)
tumorClust_downsampled <- RunUMAP(tumorClust_downsampled, 
                     reduction = "pca", 
                     dims = 1:10,
                     n.epochs = 500,
                     min.dist = 0.1,
                     seed.use = 28)

# plot UMAP
DimPlot(tumorClust_downsampled,
        reduction = "umap")

DoHeatmap(tumorClust_downsampled, features = gene.list.RNA)
DotPlot(tumorClust_downsampled, features = gene.list.RNA, cols = c("gray","darkviolet")) + coord_flip()
Clustered_DotPlot(tumorClust_downsampled, features = gene.list.RNA)


saveRDS(tumorClust_downsampled, 
        file = "/home/rstudio/mnt_out/patel2024/Rds/patel2024_downsampled_tumor_reclustering_with_IMC_genes.Rds")

```

```{r tumor_cluster}
# normalize and scale count data
tumorClust <- tumorClust %>%
    NormalizeData() %>%
    ScaleData() 

# principal component analysis
tumorClust <- RunPCA(tumorClust, 
                        assay = "RNA", 
                        npcs = 50,
                        features = gene.list.RNA)

# PCA plot
PCAPlot(object = tumorClust)

# The following 18 features requested have not been scaled 
# (running reduction without them): CD14, CD34, CD3E, CD4, CD8A, FOXP3, FUT4, 
# GATA3, ITGAM, MME, MPO, MS4A1, NCAM1, PDCD1, SOX10, CD274, ST8SIA1, B4GALNT1

# QC on how many PCs to use downstream - we want to explain 80-90% of the variance 


# find closest cells based on gene expression of the IMC marker genes
tumorClust <- FindNeighbors(tumorClust, 
                           features = gene.list.RNA,
                           reduction = "pca", 
                           dims = 1:10,
                           k.param = 100) #tested 200,500

# find clusters based on nearest neighbours
tumorClust <- FindClusters(tumorClust, 
                          resolution = 0.8, #tested 0.2, 0.4, 0.8
                          random.seed = 28,
                          group.singletons = FALSE)


# QC of UMAP parameters
library(umap)

# Tested prameters
min.dist.values <- c(0.01, 0.1, 0.3, 0.5)
dims.values <- c(10, 20, 30, 50)

# Loop over parameter combinations and visualize
for (min.dist in min.dist.values) {
  for (dims in dims.values) {
    umap_result <- RunUMAP(tumorClust, reduction = "pca", dims = 1:dims,
                           n.epochs = 500, min.dist = min.dist, 
                           features = gene.list.RNA, seed.use = 28)
    plot(umap_result, main = paste("min.dist =", min.dist, ", dims =", dims))
  }
}

# Dimension reductions with Uniform Manifold Approximation and Projection (UMAP)
tumorClust <- RunUMAP(tumorClust, 
                     reduction = "pca", 
                     n.epochs = 500,  # change to 1000, 5000 
                     min.dist = 0.1,
                     features = gene.list,
                     seed.use = 28)

# plot UMAP
DimPlot(tumorClust,
        reduction = "umap")

DoHeatmap(tumorClust, features = gene.list.RNA)
DotPlot(tumorClust, features = gene.list.RNA, cols = c("gray","darkviolet")) + coord_flip()
Clustered_DotPlot(tumorClust, features = gene.list.RNA)


saveRDS(tumorClust, file = paste(params$out,"Rds/patel2024_tumor_reclustering_with_IMC_genes.Rds"))
```

######################## 10.06.2024 #########################

# ___Label transfer from scIMC to scRNA___


```{r}
TC_clusters <- c("CD24+ marker-lo TC",
                "early SYM-like TC",
                "bridge-like TC",
                "GATA3hi TC",
                "CD24- marker-lo TC",
                "CHGAhi TC",
                "Ki67hi TC",
                "CXCR4hi TC",
                "CD44+ mes-like TC",
                "GD2lo TC")
```



```{r label transfer - tumor clusters only}
library(BiocParallel)
library(SingleR)
library(tidyr)

rna_mat <- GetAssayData(tumorClust_downsampled, assay = "RNA", slot = "data")

rna_clusters <- tumorClust_downsampled$liger_clusters
common_genes_rna <- rownames(rna_mat) %in% gene.list

rna_mat <- GetAssayData(tumorClust_downsampled, assay = "RNA", slot = "data")[common_genes_rna,]

# Take only the PT samples 
imc_PT <- subset(imc_data, tissue == "PT")
imc_PT_TC <- subset(imc_data, celltype == TC_clusters)
imc_PT_TC  # 32 proteins/genes and 27341 cells

# using all IMC clusters from PT samples
protein_mat <- GetAssayData(imc_PT, assay = "originalexp", slot = "counts")
common_genes_protein <- rownames(protein_mat) %in% rownames(rna_mat)
protein_mat <- GetAssayData(imc_PT, assay = "originalexp", slot = "counts")[common_genes_protein,]
imc_clusters <- imc_PT@meta.data$celltype
predicted_labels <- SingleR(test = rna_mat, # Transfer labels to RNA data
                            ref = protein_mat, # Original cell types from IMC data
                            labels = imc_clusters, # Cell types defined by IMC data
                            BPPARAM = MulticoreParam(workers=20, RNGseed=20231030)) 

saveRDS(predicted_labels,
        file = paste0(params$out,"Rds/20240610_predicted_labels_IMC-PT_to_scRNA-PT_Patel2024.rds"))

# using tumor IMC clusters from PT samples
protein_mat <- GetAssayData(imc_PT_TC, assay = "originalexp", slot = "counts")
common_genes_protein <- rownames(protein_mat) %in% rownames(rna_mat)
protein_mat <- GetAssayData(imc_PT_TC, assay = "originalexp", slot = "counts")[common_genes_protein,]
imc_clusters_TC <- imc_PT_TC@meta.data$celltype
predicted_labels_TC <- SingleR(test = rna_mat, # Transfer labels to RNA data
                            ref = protein_mat, # Original cell types from IMC data
                            labels = imc_clusters_TC, # Cell types defined by IMC data
                            BPPARAM = MulticoreParam(workers=20, RNGseed=20231030)) 

saveRDS(predicted_labels_TC,
        file = paste0(params$out,"Rds/20240610_predicted_labels_IMC-PT-TC_to_scRNA-PT-TC_Patel2024.rds"))

```

```{r heatmap of predicted labels}
library(viridis)
library(RColorBrewer)
library(ComplexHeatmap)
#library(magick)
library(scater)
library(patchwork)
library(cowplot)

set.seed(2805)

predicted_labels <- 
  readRDS(file = paste0(params$out,"Rds/20240610_predicted_labels_IMC-PT_to_scRNA-PT_Patel2024.rds"))

predicted_labels_TC <- 
  readRDS(file = paste0(params$out,"Rds/20240610_predicted_labels_IMC-PT-TC_to_scRNA-PT-TC_Patel2024.rds"))

plotScoreHeatmap(predicted_labels_TC[sample(seq_len(nrow(predicted_labels_TC)), 2000),], 
                 annotation_legend=T, treeheight_row=0) 

pdf(paste0(params$out,"figures/labeltransfer_MC-PT_to_scRNA-PT_Patel2024_score_heatmap.pdf"))
plotScoreHeatmap(predicted_labels_TC[sample(seq_len(nrow(predicted_labels_TC)), 2000),], 
                 annotation_legend=T, treeheight_row=0) 
dev.off()

tumorClust_downsampled@meta.data$original_celltype <- tumorClust_downsampled@meta.data$liger_clusters
true_label <- tumorClust_downsampled@meta.data$liger_clusters

# all clusters
predicted_label <- predicted_labels$labels
tumorClust_downsampled@meta.data$predicted_celltype <- predicted_label

# TC only
predicted_label_TC <- predicted_labels_TC$labels
tumorClust_downsampled@meta.data$predicted_celltype_TC <- predicted_label_TC

# Idents(tumorClust_downsampled) <- tumorClust_downsampled@meta.data$predicted_celltype_TC


n <- length(unique(predicted_label_TC))

table(predicted_label_TC)

rna_mat_scaled <-  t(scale(t(rna_mat)))

qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector_74 = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
col_vector_433 = colors()[grep('gr(a|e)y', grDevices::colors(), invert = T)]

col_predicted_labels = setNames(sample(col_vector_433, length(unique(predicted_label_TC))), unique(predicted_label_TC))
col_true_labels <- tumorClust_downsampled@meta.data$liger_clusters

column_ha = HeatmapAnnotation(predicted_label = predicted_label_TC[rna_mat_scaled],
                              true_cluster = true_label[rna_mat_scaled],
                              col=list(predicted_label=col_predicted_labels, 
                                       true_label=col_true_labels
                                       ), 
                              show_legend=T)

quantile(rna_mat_scaled, c(0.1, 0.95))

## make the white color map to 0. the green map to highest and the purle map to the lowest
col_fun = circlize::colorRamp2(c(-0.5, 0, 0.5, 1, 1.5), 
                               c("purple", "white", "lightgreen", "green4","darkgreen"))

tumorClust_downsampled <- SetIdent(tumorClust_downsampled, value = "predicted_celltype")

x <- Heatmap(rna_mat_scaled, name = "IMC markers gene expression",  
             #column_split = factor(predicted_label),
             cluster_columns = TRUE,
             show_column_dend = TRUE,
             cluster_column_slices = TRUE,
             column_title_gp = gpar(fontsize = 8),
             column_gap = unit(0.5, "mm"),
             cluster_rows = TRUE,
             show_row_dend = TRUE,
             col = col_fun,
             row_names_gp = gpar(fontsize = 4),
             column_title_rot = 90,
             #top_annotation = HeatmapAnnotation(foo = anno_block(gp = gpar(fill = scales::hue_pal()(9)))),
             show_column_names = FALSE,
             use_raster = TRUE,
             raster_quality = 4)

pdf(paste0(params$out,"figures/labeltransfer_MC-PT_to_scRNA-PT_Patel2024_expression_heatmap.pdf"))
x
dev.off()


DimPlot(tumorClust_downsampled,cols = COLOR_CODE_IMC)


DotPlot(tumorClust_downsampled, features = gene.list) + RotatedAxis()

pdf(paste0(params$out,"figures/labeltransfer_MC-PT_to_scRNA-PT_Patel2024_marker_expression_dotplot.pdf"))
DotPlot(tumorClust_downsampled, features = gene.list) + RotatedAxis()
dev.off()


tumor_markers <- c("ELAVL4", "CXCR4", "MME", "CD74", "HLA-A", "CD274", "GATA3", 
                   "VIM", "MKI67", "CD44", "NCAM1", "ST8SIA1", "PRPH", "LUM", 
                   "CHGA", "SOX10", "S100B")

pdf(paste0(params$out,"figures/labeltransfer_MC-PT_to_scRNA-PT_Patel2024_tumor_marker_expression_dotplot.pdf"))
DotPlot(tumorClust_downsampled, features = tumor_markers) + RotatedAxis()
dev.off()

DoHeatmap(tumorClust_downsampled,
          features = gene.list, 
          size = 3, 
          slot = "counts")


# Plot in the UMAP published and predicted cell types (based on IMC protein data)

data <- data.frame(IMC_cell_types=predicted_label_TC, 
                   original_clusters=tumorClust_downsampled@meta.data$liger_clusters)

data$freq <- 1
data_wide <- data %>% pivot_wider(names_from = c("original_clusters"), 
                                  values_from = "freq", 
                                  values_fn = sum, 
                                  values_fill = 0)

data_long <- data_wide %>% pivot_longer(cols = !IMC_cell_types, 
                                        names_to = "original_clusters", 
                                        values_to = "count") 


IMC_cell_types <- data_wide$IMC_cell_types
data <- data_wide
data <- data[,-1]
rownames(data) <- IMC_cell_types
data <- (prop.table(as.matrix(data), margin=2))
original_clusters <- names(unlist(apply(data, 2, which.max)))
predicted_clust_1 <- data_wide[unname(unlist(apply(data, 2, which.max))), "IMC_cell_types"]
proportion_1 <- unname(unlist(apply(data, 2, max)))
predicted_clust_2 <- unname(apply(data, 2, function(x){rownames(data)[order(x, decreasing=T)][2]}))
proportion_2 <- unname(apply(data, 2, function(x){sort(x, decreasing=T)[2]}))

assignments <- as.data.frame(cbind("original_clusters" = original_clusters, 
                                   "predicted_cluster_1" = predicted_clust_1, 
                                   "proportion_1" = proportion_1, 
                                   "predicted_cluster_2" = predicted_clust_2, 
                                   "proportion_2" = proportion_2))

assignments


#rna <- SetIdent(rna, value = "predicted_celltype")


bar_plot <- ggplot(data_long, aes(fill=IMC_cell_types, y=count, x=original_clusters)) + 
  ggtitle("Barplot of predicted vs. original labels") +
  geom_bar(position="fill", stat="identity", show.legend = T) + 
  scale_fill_manual(values = COLOR_CODE_IMC) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 

bar_plot

pdf(paste0(params$out,"figures/barplot_predicted_celltypes_MC-PT_to_scRNA-PT_Patel2024_proportion.pdf"),
    height = 10,width = 20)
bar_plot
dev.off()


## UMAP colored by cell type and expression 
DimPlot(tumorClust_downsampled, 
             reduction = "umap", 
             pt.size = 0.3, 
             seed = 28, 
             label =T, 
             group.by = "predicted_celltype")



pdf(file=paste0(params$out,"figures/umap_MC-PT_to_scRNA-PT_Patel2024_labelTransfer.pdf"),
    height = 10, width = 20)
p1
dev.off()


table(tumorClust_downsampled$predicted_celltype)
table(tumorClust_downsampled$original_celltype)

compare_cell_types <- c(tumorClust_downsampled$original_celltype,
                        tumorClust_downsampled$predicted_celltype)


# Rename clusters based in singleR results
Idents(tumorClust_downsampled) <- tumorClust_downsampled$seurat_clusters
tumorClust_downsampled <- RenameIdents(object = tumorClust_downsampled, 
                                       "0" = "CD44+ mes-like TC",
                                       "1" = "CD24- marker-lo TC",
                                       "2" = "CHGAhi TC",
                                       "3" = "bridge-like TC",
                                       "4" = "CD24- marker-lo TC",
                                       "5" = "Ki67hi TC",
                                       "6" = "CD24- marker-lo TC",
                                       "7" = "early SYM-like TC",
                                       "8" = "CD24- marker-lo TC",
                                       "9" = "CD24- marker-lo TC")


# Rename clusters based in singleR results
Idents(tumorClust_downsampled) <- tumorClust_downsampled$liger_clusters
tumorClust_downsampled <- RenameIdents(object = tumorClust_downsampled, 
                                       "1" = "early SYM-like TC",
                                       "2" = "CD24- marker-lo TC",
                                       "3" = "CD24- marker-lo TC",
                                       "4" = "CD24- marker-lo TC",
                                       "5" = "CD24- marker-lo TC",
                                       "6" = "CD24- marker-lo TC",
                                       "7" = "CD44+ mes-like TC",
                                       "8" = "CD44+ mes-like TC",
                                       "9" = "CD24- marker-lo TC",
                                       "10" = "CD24- marker-lo TC",
                                       "11" = "CD24- marker-lo TC",
                                       "12" = "CD24- marker-lo TC"
                                       )


DotPlot(tumorClust_downsampled, 
        features = gene.list.RNA, 
        cols = c("gray","purple"))  + RotatedAxis()

saveRDS(object = tumorClust_downsampled, 
        file="/home/rstudio/mnt_out/patel2024/Rds/Patel2024_AfterLabelTransfer.Rds")

```



------------------------------------------------------------------------

```{r}
sessionInfo()
```


