---
title: "UMAP and histograms after annotation"
output: html_document
date: '2022-12-28'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
```

In this RMD file, we will plot UMAPs and histograms on the annotated SPE.

```{r read-data}

path <- "/mnt/Multimodal_Imaging_Daria/PIPELINE/NEWEST_RESULTS"

spe <- readRDS(file.path(path,"results/Manuscript/spe_clustered.rds"))
```


Then we will generate a matrix of the expression data to be able to run the plots below. 

```{r generate-dataframes, message=FALSE, fig.height=7}
library(SpatialExperiment)

mat_lq <- assay(spe, "exprs")[!rowData(spe)$use_channel,]
mat_hq <- assay(spe, "exprs")[rowData(spe)$use_channel,]
mat_morph <- rbind(mat_hq, "area"=spe$area_clipped, "solidity"=spe$solidity_clipped, "aspect_ratio"=spe$aspect_ratio_clipped)
```

Then we will plot a heatmap of 2000 randomly selected cells. 

```{r single-cell-heatmap-clusters, message=FALSE, fig.height=17}
library(ComplexHeatmap)
library(viridis)
library(circlize)
library(RColorBrewer)
library(paletteer)
set.seed(20230521)

mat <- mat_hq
celltype <- spe$celltype
tissue <- spe$tissue

col_marker = colorRamp2(c(0, 2.5, 0.5), c("blue", "white", "red"))
col_tissue = c("PT" = "deepskyblue", "BM" = "orange")

rand_cells <- sample(seq_len(ncol(mat)), 2000)


area = mat_morph["area",]
solidity = mat_morph["solidity",]
aspect_ratio = mat_morph["aspect_ratio",]

clPARP = mat_lq["clPARP_Yb176_mean",]

col_area = colorRamp2(c(min(area[rand_cells]),min(area[rand_cells])+(max(area[rand_cells])-min(area[rand_cells]))/2, max(area[rand_cells])), c("blue", "white", "red"))
col_solidity = colorRamp2(c(min(solidity[rand_cells]),min(solidity[rand_cells])+(max(solidity[rand_cells])-min(solidity[rand_cells]))/2, max(solidity[rand_cells])), c("blue", "white", "red"))
col_aspect_ratio = colorRamp2(c(min(aspect_ratio[rand_cells]),min(aspect_ratio[rand_cells])+(max(aspect_ratio[rand_cells])-min(aspect_ratio[rand_cells]))/2, max(aspect_ratio[rand_cells])), c("blue", "white", "red"))

col_clPARP = colorRamp2(c(min(clPARP[rand_cells]),min(clPARP[rand_cells])+(max(clPARP[rand_cells])-min(clPARP[rand_cells]))/2, max(clPARP[rand_cells])), c("blue", "white", "red"))

column_ha = HeatmapAnnotation(celltype = celltype[rand_cells],
                              tissue = tissue[rand_cells],
                              area = area[rand_cells],
                              solidity = solidity[rand_cells],
                              aspect_ratio = aspect_ratio[rand_cells],
                              #clPARP = clPARP[rand_cells],
                              col=list(celltype=metadata(spe)$color_vectors$col_celltype, 
                                       tissue=col_tissue,
                                       area = col_area,
                                       solidity = col_solidity,
                                       aspect_ratio = col_aspect_ratio#,
                                       #clPARP = col_clPARP
                                       ), 
                              show_legend=T)


Heatmap(mat[,rand_cells], 
        column_title = "Single-cell heatmap with 2000 randomly selected cells",
        top_annotation = column_ha, 
        cluster_columns=F,
        column_order= order(celltype[rand_cells], decreasing=F),
        show_column_dend = T,
        show_column_names=F,
        col=viridis(100)
        )
```

Now we will plot these heatmaps per cluster.

```{r single-cell-heatmap-clusters, message=FALSE, fig.height=17}

set.seed(20230521)

celltype <- spe$celltype

for(i in 1: length(unique(celltype))){

  c <- unique(celltype)[i]

  mat_c <- mat_hq[,celltype==c]
  ct <- celltype[celltype==c]
  tis <- spe$tissue[celltype==c]
  
  col_marker = colorRamp2(c(0, 2.5, 0.5), c("blue", "white", "red"))
  col_tissue = c("PT" = "deepskyblue", "BM" = "orange")
  
  if (length(ct) < 8000){len=length(ct)}else{len=8000}

  rand_cells <- sample(seq_len(ncol(mat_c)), len)
  
  
  area_c = mat_morph["area",celltype==c]
  solidity_c = mat_morph["solidity",celltype==c]
  aspect_ratio_c = mat_morph["aspect_ratio",celltype==c]
  
  clPARP_c = mat_lq["clPARP_Yb176_mean",celltype==c]
  
  col_area_c = colorRamp2(c(min(area_c[rand_cells]),min(area_c[rand_cells])+(max(area_c[rand_cells])-min(area_c[rand_cells]))/2, max(area_c[rand_cells])), c("blue", "white", "red"))
  col_solidity_c = colorRamp2(c(min(solidity_c[rand_cells]),min(solidity_c[rand_cells])+(max(solidity_c[rand_cells])-min(solidity_c[rand_cells]))/2, max(solidity_c[rand_cells])), c("blue", "white", "red"))
  col_aspect_ratio_c = colorRamp2(c(min(aspect_ratio_c[rand_cells]),min(aspect_ratio_c[rand_cells])+(max(aspect_ratio_c[rand_cells])-min(aspect_ratio_c[rand_cells]))/2, max(aspect_ratio_c[rand_cells])), c("blue", "white", "red"))
  
  col_clPARP_c = colorRamp2(c(min(clPARP_c[rand_cells]),min(clPARP_c[rand_cells])+(max(clPARP_c[rand_cells])-min(clPARP_c[rand_cells]))/2, max(clPARP_c[rand_cells])), c("blue", "white", "red"))
  
  column_ha = HeatmapAnnotation(celltype = ct[rand_cells],
                                tissue = tis[rand_cells],
                                area = area_c[rand_cells],
                                solidity = solidity_c[rand_cells],
                                aspect_ratio = aspect_ratio_c[rand_cells],
                                clPARP = clPARP_c[rand_cells],
                                col=list(celltype=metadata(spe)$color_vector$col_celltype, 
                                         tissue=col_tissue,
                                         area = col_area_c,
                                         solidity = col_solidity_c,
                                         aspect_ratio = col_aspect_ratio_c,
                                         clPARP = col_clPARP_c
                                         ), 
                                show_legend=T)
  
  png(file=file.path(path, paste0("results/Manuscript/1_Figure/", c , ".tiff")),width=18,height=8,units="in", res=1200)
  
  ht<- Heatmap(mat_c[,rand_cells], 
          column_title = "Single-cell heatmap with 8000 randomly selected cells",
          top_annotation = column_ha, 
          cluster_columns=F,
          column_order= order(tis[rand_cells], decreasing=F),
          show_column_dend = T,
          show_column_names=F,
          col=viridis(100)
          )
  draw(ht)
  
  dev.off()
}

```

Next we will look at the single cell heatmap ordered by the identified phenograph clusters and sample the same number of cells from BM and PT samples. 

```{r single-cell-heatmap-clusters, message=FALSE, fig.height=7, fig.width=14}
library(ComplexHeatmap)
library(viridis)
library(circlize)
library(RColorBrewer)
library(paletteer)

set.seed(20230618) 

col_marker = colorRamp2(c(0, 2.5, 0.5), c("blue", "white", "red"))
col_tissue = c("PT" = "deepskyblue", "BM" = "orange")

#Get same number of PT and BM cells
pt_index <- which(tissue=="PT")
bm_index <- which(tissue=="BM")
rand_cells_pt <- sample(seq_len(length(pt_index)), 4000)
rand_cells_bm <- sample(seq_len(length(bm_index)), 4000)
pt_index_rand <- pt_index[rand_cells_pt]
bm_index_rand <- bm_index[rand_cells_bm]
rand_cells <- c(pt_index_rand, bm_index_rand)

col_area = colorRamp2(c(min(area[rand_cells]),min(area[rand_cells])+(max(area[rand_cells])-min(area[rand_cells]))/2, max(area[rand_cells])), c("blue", "white", "red"))
col_solidity = colorRamp2(c(min(solidity[rand_cells]),min(solidity[rand_cells])+(max(solidity[rand_cells])-min(solidity[rand_cells]))/2, max(solidity[rand_cells])), c("blue", "white", "red"))
col_aspect_ratio = colorRamp2(c(min(aspect_ratio[rand_cells]),min(aspect_ratio[rand_cells])+(max(aspect_ratio[rand_cells])-min(aspect_ratio[rand_cells]))/2, max(aspect_ratio[rand_cells])), c("blue", "white", "red"))

#col_clPARP = colorRamp2(c(min(clPARP[rand_cells]),min(clPARP[rand_cells])+(max(clPARP[rand_cells])-min(clPARP[rand_cells]))/2, max(clPARP[rand_cells])), c("blue", "white", "red"))

column_ha = HeatmapAnnotation(cluster = celltype[rand_cells],
                              tissue = tissue[rand_cells],
                              area = area[rand_cells],
                              solidity = solidity[rand_cells],
                              aspect_ratio = aspect_ratio[rand_cells],
                              #clPARP = clPARP[rand_cells],
                              col=list(cluster=metadata(spe)$color_vector$col_celltype, 
                                       tissue=metadata(spe)$color_vectors$col_tissue,
                                       area = col_area,
                                       solidity = col_solidity,
                                       aspect_ratio = col_aspect_ratio#,
                                       #clPARP = col_clPARP
                                       ), 
                              show_legend=T)


Heatmap(mat[,rand_cells], 
        column_title = "Single-cell heatmap showing 4000 random cells per tissue type",
        top_annotation = column_ha, 
        cluster_columns=F,
        row_labels=sapply(strsplit(rownames(mat), "_"), "[[", 1),
        column_order= order(celltype[rand_cells], decreasing=F),
        show_column_dend = T,
        show_column_names=F,
        col=viridis(100)
        )
```
Next we will look at the single cell heatmap ordered by the identified phenograph clusters and sample the same number of cells from BM and PT samples per cluster. 

```{r single-cell-heatmap-clusters, message=FALSE, fig.height=17}

set.seed(20230517)

n <- length(unique(celltype))

count <- 45
rand_cells <- c()

#Get same number of PT and BM cells per cluster
for (i in 1: length(unique(celltype))){
  pt_index <- which(tissue=="PT" & celltype == unique(celltype)[i])
  bm_index <- which(tissue=="BM" & celltype == unique(celltype)[i])
  if (length(pt_index) < count){
    pt_diff <- count-length(pt_index)
    print(paste("PT_diff: ", pt_diff, " for cluster ", unique(celltype)[i]))
    rand_index_pt <- sample(pt_index, length(pt_index))
    rand_index_bm <- sample(bm_index, (count + pt_diff))
  } else if (length(bm_index) < count){
    bm_diff <- count-length(bm_index)
    rand_index_pt <- sample(pt_index, (count + bm_diff))
    rand_index_bm <- sample(bm_index, length(bm_index))
  } else {
      rand_index_pt <- sample(pt_index, count)
      rand_index_bm <- sample(bm_index, count)
  }

  rand_cells <- c(rand_cells, rand_index_pt, rand_index_bm)
}

col_area = colorRamp2(c(min(area[rand_cells]),min(area[rand_cells])+(max(area[rand_cells])-min(area[rand_cells]))/2, max(area[rand_cells])), c("blue", "white", "red"))
col_solidity = colorRamp2(c(min(solidity[rand_cells]),min(solidity[rand_cells])+(max(solidity[rand_cells])-min(solidity[rand_cells]))/2, max(solidity[rand_cells])), c("blue", "white", "red"))
col_aspect_ratio = colorRamp2(c(min(aspect_ratio[rand_cells]),min(aspect_ratio[rand_cells])+(max(aspect_ratio[rand_cells])-min(aspect_ratio[rand_cells]))/2, max(aspect_ratio[rand_cells])), c("blue", "white", "red"))

col_marker = colorRamp2(c(0, 2.5, 0.5), c("blue", "white", "red"))
col_tissue = c("PT" = "deepskyblue", "BM" = "orange")

#mat <- t(mat)
                                       
                                       
column_ha = HeatmapAnnotation(cluster = celltype[rand_cells],
                              tissue = tissue[rand_cells],
                              area = area[rand_cells],
                              solidity = solidity[rand_cells],
                              aspect_ratio = aspect_ratio[rand_cells],
                              col=list(cluster=metadata(spe)$color_vector$col_celltype, 
                                       tissue=col_tissue,
                                       area = col_area,
                                       solidity = col_solidity,
                                       aspect_ratio = col_aspect_ratio
                                       ), 
                              show_legend=T)


Heatmap(mat[,rand_cells], 
        column_title = "Single-cell heatmap same number of cells from each tissue per cluster",
        top_annotation = column_ha, 
        cluster_columns=F,
        cluster_rows=F,
        column_order= order(celltype[rand_cells], decreasing=F),
        show_column_dend = T,
        show_column_names=F,
        col=viridis(100)
        )
```

Next we will look at the single cell heatmap ordered by the identified phenograph clusters and sample the same number of cells from each cluster

```{r single-cell-heatmap-clusters, message=FALSE, fig.height=17}

set.seed(20230517)

n <- length(unique(celltype))

count <- 90
rand_cells <- c()

#Get same number of PT and BM cells per cluster
for (i in 1: length(unique(celltype))){
  idx <- which(celltype == unique(celltype)[i])
  if(length(idx)<count)
    {rand_idx <- sample(idx, length(idx))}
  else
    {rand_idx <- sample(idx, count)}
  
  rand_cells <- c(rand_cells, rand_idx)
}

col_area = colorRamp2(c(min(area[rand_cells]),min(area[rand_cells])+(max(area[rand_cells])-min(area[rand_cells]))/2, max(area[rand_cells])), c("blue", "white", "red"))
col_solidity = colorRamp2(c(min(solidity[rand_cells]),min(solidity[rand_cells])+(max(solidity[rand_cells])-min(solidity[rand_cells]))/2, max(solidity[rand_cells])), c("blue", "white", "red"))
col_aspect_ratio = colorRamp2(c(min(aspect_ratio[rand_cells]),min(aspect_ratio[rand_cells])+(max(aspect_ratio[rand_cells])-min(aspect_ratio[rand_cells]))/2, max(aspect_ratio[rand_cells])), c("blue", "white", "red"))

col_marker = colorRamp2(c(0, 2.5, 0.5), c("blue", "white", "red"))
col_tissue = c("PT" = "deepskyblue", "BM" = "orange")

#mat <- t(mat)

column_ha = HeatmapAnnotation(metacluster=spe$metacluster[rand_cells],
                              cluster = celltype[rand_cells],
                              tissue = tissue[rand_cells],
                              area = area[rand_cells],
                              solidity = solidity[rand_cells],
                              aspect_ratio = aspect_ratio[rand_cells],
                              col=list(metacluster=metadata(spe)$color_vectors$col_metacluster,
                                       cluster=metadata(spe)$color_vectors$col_celltype, 
                                       tissue=col_tissue,
                                       area = col_area,
                                       solidity = col_solidity,
                                       aspect_ratio = col_aspect_ratio
                                       ), 
                              annotation_legend_param = list(cluster = list(at = unique(celltype[rand_cells][order(as.integer(sub('_.*', '', celltype[rand_cells])), decreasing=F)]))), #delete later
                              show_legend=T)


Heatmap(mat[,rand_cells], 
        column_title = "Single-cell heatmap showing clusters of equal number of randomly selected cells",
        top_annotation = column_ha, 
        cluster_columns=F,
        cluster_rows=F,
        #column_order= order(celltype[rand_cells], decreasing=F),
        column_order = order(as.integer(sub('_.*', '', celltype[rand_cells])), decreasing=F),
        show_column_dend = T,
        show_column_names=F,
        col=viridis(100)
        )
```

Next, we will look at the mean marker expression per cluster. The next two code blocks can be also run for tissue types (PT/BM) separately. 

```{r ingle-cell-heatmap-clusters-mean, message=FALSE, fig.width=16, fig.height=12}
library(scuttle)

set.seed(20230135)

#calculate mean heatmap over all cells
matrix_hq <- mat_hq
matrix_lq <- mat_lq
morph_matrix <- mat_morph
clust <- spe$celltype

#calculate heatmap over tissue-derived cells
# tissue <- "PT"
# matrix_hq <- mat_hq[,spe$tissue==tissue]
# matrix_lq <- mat_lq[,spe$tissue==tissue]
# morph_matrix <- mat_morph[,spe$tissue==tissue]
# clust <- celltype[spe$tissue==tissue]

#Mean of high quality markers per cluster
mean <- aggregate(t(mat_hq), list(clust), mean)
mean_t <- t(mean[,2:ncol(mean)])
mean_t <- t(apply(mean_t, 1, RESCALE, to=0:1))
colnames(mean_t) <- mean[,"Group.1"]

#Mean of morphological features per cluster
mean_morph <- aggregate(t(morph_matrix), list(clust), mean)
area_mean = mean_morph[,"area"]
solidity_mean = mean_morph[,"solidity"]
aspect_ratio_mean = mean_morph[,"aspect_ratio"]

#Mean of low quality markers per cluster
mean_lq <- aggregate(t(matrix_lq), list(clust), mean)
clPARP_mean = mean_lq[,"clPARP_Yb176_mean"]
FN1_mean = mean_lq[,"Fibronectin_La139_mean"]

#Metacluster
mean_metaclust <-aggregate(spe$metacluster, list(clust), max)
mean_metacluster <- mean_metaclust$x

col_mean_area = colorRamp2(c(min(area_mean),min(area_mean)+(max(area_mean)-min(area_mean))/2, max(area_mean)), c("blue", "white", "red"))
col_mean_solidity = colorRamp2(c(min(solidity_mean),min(solidity_mean)+(max(solidity_mean)-min(solidity_mean))/2, max(solidity_mean)), c("blue", "white", "red"))
col_mean_aspect_ratio = colorRamp2(c(min(aspect_ratio_mean),min(aspect_ratio_mean)+(max(aspect_ratio_mean)-min(aspect_ratio_mean))/2, max(aspect_ratio_mean)), c("blue", "white", "red"))
col_mean_clPARP = colorRamp2(c(min(clPARP_mean),min(clPARP_mean)+(max(clPARP_mean)-min(clPARP_mean))/2, max(clPARP_mean)), c("blue", "white", "red"))
col_mean_FN1 = colorRamp2(c(min(FN1_mean),min(FN1_mean)+(max(FN1_mean)-min(FN1_mean))/2, max(FN1_mean)), c("blue", "white", "red"))

column_ha = HeatmapAnnotation(metacluster=mean_metacluster,
                              celltype = mean[,"Group.1"], 
                              area=area_mean,
                              solidity = solidity_mean,
                              aspect_ratio = aspect_ratio_mean,
                              # clPARP = clPARP_mean,
                              # FN1 = FN1_mean,
                              ncells=anno_barplot(as.data.frame(table(clust))[,"Freq"], height = unit(0.6, "cm")),
                              col=list(metacluster=metadata(spe)$color_vectors$col_metacluster,
                                       celltype=metadata(spe)$color_vectors$col_celltype, 
                                       area=col_mean_area,
                                       solidity=col_mean_solidity,
                                       aspect_ratio=col_mean_aspect_ratio#,
                                       # clPARP = col_mean_clPARP,
                                       # FN1 = col_mean_FN1
                                       ), 
                              show_legend=F,
                             annotation_name_gp= gpar(fontsize = 7),
                             simple_anno_size = unit(0.15, "cm"))


h1 <- Heatmap(mean_t, 
        #column_title = "Heatmap showing mean per cluster - unscaled",
        top_annotation = column_ha, 
        cluster_columns=T,
        clustering_distance_columns="pearson", #use euclidean, kendall or pearson
        clustering_method_columns="single",
        cluster_rows=T,
        row_labels=sapply(strsplit(rownames(mean_t), "_"), "[[", 1),
        show_column_dend = F,
        show_row_dend = F,
        show_column_names=T,
        show_row_names=T,
        col=viridis(100),
        column_names_gp = grid::gpar(fontsize = 6),
        row_names_gp = grid::gpar(fontsize = 7),
        #column_dend_height = unit(4, "cm"),
        column_names_rot = 45
        )

# tiff(filename = "/mnt/Multimodal_Imaging_Daria/PIPELINE/NEWEST_RESULTS/results/poster/heatmap_mean-per-cluster_flashtalk.tiff",
#      width = 250, height = 160, units = "mm", 
#      res=600)
# h1
# dev.off()

library(tidyHeatmap)
save_pdf(
  h1,
  "/mnt/Multimodal_Imaging_Daria/PIPELINE/NEWEST_RESULTS/results/poster/heatmap_mean-per-cluster_legend.pdf",
  width = 208,
  height = 110,
  units = c("mm")
)

```
We can see that cluster 38 are the sticky cells, which are positive for everything. We will exclude this cluster and re-generate the heatmap.

```{r ingle-cell-heatmap-clusters-mean, message=FALSE, fig.height=20, fig.width=18}
library(scuttle)
library(bruceR)

sticky_cluster <- "other"

# Feature-wise scaling
mean_t_wo_sticky <- mean_t[,colnames(mean_t)!=sticky_cluster]
mean_t_wo_sticky <- t(apply(mean_t_wo_sticky, 1, RESCALE, to=0:1))
clusters_wo_sticky <- spe$celltype[spe$celltype!=sticky_cluster]

area_mean_wo_sticky <- area_mean[colnames(mean_t)!=sticky_cluster]
solidity_mean_wo_sticky <- solidity_mean[colnames(mean_t)!=sticky_cluster]
aspect_ratio_mean_wo_sticky <- aspect_ratio_mean[colnames(mean_t)!=sticky_cluster]
clPARP_mean_wo_sticky <- clPARP_mean[colnames(mean_t)!=sticky_cluster]


col_mean_area_wo_sticky = colorRamp2(c(min(area_mean_wo_sticky),min(area_mean_wo_sticky)+(max(area_mean_wo_sticky)-min(area_mean_wo_sticky))/2, max(area_mean_wo_sticky)), c("blue", "white", "red"))
col_mean_solidity_wo_sticky = colorRamp2(c(min(solidity_mean_wo_sticky),min(solidity_mean_wo_sticky)+(max(solidity_mean_wo_sticky)-min(solidity_mean_wo_sticky))/2, max(solidity_mean_wo_sticky)), c("blue", "white", "red"))
col_mean_aspect_ratio_wo_sticky = colorRamp2(c(min(aspect_ratio_mean_wo_sticky),min(aspect_ratio_mean_wo_sticky)+(max(aspect_ratio_mean_wo_sticky)-min(aspect_ratio_mean_wo_sticky))/2, max(aspect_ratio_mean_wo_sticky)), c("blue", "white", "red"))
col_mean_clPARP_wo_sticky = colorRamp2(c(min(clPARP_mean_wo_sticky),min(clPARP_mean_wo_sticky)+(max(clPARP_mean_wo_sticky)-min(clPARP_mean_wo_sticky))/2, max(clPARP_mean_wo_sticky)), c("blue", "white", "red"))

col_celltype <- metadata(spe)$color_vectors$col_celltype
col_clusters_wo_sticky <- col_celltype[names(col_celltype)!=sticky_cluster]

mean_metacluster_wo_sticky <- mean_metaclust[colnames(mean_t)!=sticky_cluster,]$x

column_ha = HeatmapAnnotation(metacluster=mean_metacluster_wo_sticky,
                              celltype = colnames(mean_t_wo_sticky),
                              area=area_mean_wo_sticky,
                              solidity=solidity_mean_wo_sticky,
                              aspect_ratio=aspect_ratio_mean_wo_sticky,
                              clPARP=clPARP_mean_wo_sticky,
                              ncells=anno_barplot(as.data.frame(table(clusters_wo_sticky))[,"Freq"], height = unit(3, "cm")),
                              col=list(metacluster=metadata(spe)$color_vectors$col_metacluster, 
                                       celltype=col_clusters_wo_sticky,
                                       area=col_mean_area_wo_sticky,
                                       solidity=col_mean_solidity_wo_sticky,
                                       aspect_ratio=col_mean_aspect_ratio_wo_sticky,
                                       clPARP=col_mean_clPARP_wo_sticky
                                       ), 
                              show_legend=T)

Heatmap(mean_t_wo_sticky, 
        column_title = "Heatmap showing mean per cluster - unscaled",
        top_annotation = column_ha, 
        cluster_columns=T,
        clustering_distance_columns="pearson", #use euclidean, kendal or pearson
        clustering_method_columns="single",
        cluster_rows=T,
        row_labels=sapply(strsplit(rownames(mean_t_wo_sticky), "_"), "[[", 1),
        show_column_dend = T,
        show_column_names=T,
        show_row_names=T,
        col=viridis(100),
        column_dend_height = unit(8, "cm")
        )



```
```{r dot-plot, message=FALSE, fig.height=12, fig.width=10}
library(sccore)
library(parallel)
library(viridis)
library(dplyr)

cores <- detectCores()
rand_cells <- sample(seq_len(ncol(mat)), 200000)

ct <- spe$celltype[rand_cells] %>% factor() %>% setNames(colnames(spe)[rand_cells])
d <- dotPlot(markers = rownames(mat), count.matrix = t(mat[,rand_cells]), cell.groups = ct, n.cores=cores-2) + scale_color_viridis_c()
df <- d$data
```
The problem above is that the internal scaling can't be turned off. So we will try to plot a dotplot with complex heatmap according to [this](https://divingintogeneticsandgenomics.com/post/clustered-dotplot-for-single-cell-rnaseq/).

```{r dot-plot-complex-heatmap, message=FALSE, fig.height=14, fig.width=16}
library(tidyr)
library(tidyHeatmap)
library(bruceR)

#Mean expression per cluster

exp_mat_ <- aggregate(t(mat), list(spe$celltype), mean)
exp_mat <- exp_mat_[,2:ncol(exp_mat_)]
exp_mat <- apply(exp_mat, 2, RESCALE, to=0:1)
rownames(exp_mat) <- exp_mat_[,"Group.1"]
exp_mat <- t(exp_mat)
head(exp_mat)

#Percentage of cells per cluster expressing a marker (>0)
tmp <- mat
tmp[tmp > unname(quantile(exp_mat[exp_mat>0], c(0.2)))] <- 1
tmp[tmp <= unname(quantile(exp_mat[exp_mat>0], c(0.2)))] <- 0
tmp <- aggregate(t(tmp), list(spe$celltype), sum)
tmp2 <- tmp[,2:ncol(tmp)]
rownames(tmp2) <- tmp[,"Group.1"]
percent_mat <- tmp2 / unname(table(spe$celltype))
percent_mat <- t(percent_mat)
percent_mat[percent_mat < 0.1] <- 0.1 #setting min values higher to make points in heatmap visible
head(percent_mat)

#Annotation bars
bm_count <- table(spe[,spe$tissue=="BM"]$celltype)
bm_count[setdiff(colnames(exp_mat), names(bm_count))] <- 0
bm_count <- bm_count[colnames(exp_mat)]
pt_count <- table(spe[,spe$tissue=="PT"]$celltype)
pt_count[setdiff(colnames(exp_mat), names(pt_count))] <- 0
pt_count <- pt_count[colnames(exp_mat)]

mean_metaclust <-aggregate(spe$metacluster, list(spe$celltype), max)

mean_morph <- aggregate(t(mat_morph), list(spe$celltype), mean)
area_mean = mean_morph[,"area"]
solidity_mean = mean_morph[,"solidity"]
col_mean_area = colorRamp2(c(min(area_mean),min(area_mean)+(max(area_mean)-min(area_mean))/2, max(area_mean)), c("blue", "white", "red"))
col_mean_solidity = colorRamp2(c(min(solidity_mean),min(solidity_mean)+(max(solidity_mean)-min(solidity_mean))/2, max(solidity_mean)), c("blue", "white", "red"))

column_ha = HeatmapAnnotation(metacluster=mean_metaclust$x,
                              #celltype = colnames(exp_mat),
                              roundness = solidity_mean,
                              #area = area_mean,
                              #ncells_overall=anno_barplot(as.data.frame(table(spe$celltype))[,"Freq"], add_numbers=T, height = unit(2.2, "cm")),
                              ncells_PT=anno_barplot(unname(pt_count), add_numbers=F, height = unit(0.7, "cm"), gp = gpar(fill = "brown")),
                              ncells_bm=anno_barplot(unname(bm_count), add_numbers=F, height = unit(0.7, "cm"), gp = gpar(fill = "aquamarine4")),
                              #ncells = anno_barplot(matrix(nc=2, c(unname(pt_count), unname(bm_count))), beside=T, attach=T, height = unit(2.2, "cm")),
                              col=list(metacluster=metadata(spe)$color_vectors$col_metacluster, 
                                       #celltype=metadata(spe)$color_vectors$col_celltype
                                       roundness = col_mean_solidity
                                       #area = col_mean_area
                                       ), 
                              #annotation_legend_param = list(celltype = list(
                       	      #at = colnames(exp_mat)[order(mean_metaclust$x)])),
                              show_legend=T)

#Function to set dot size based on percentage expressed
cell_fun = function(j, i, x, y, w, h, fill){
          grid.rect(x = x, y = y, width = w, height = h, 
                    gp = gpar(col = NA, fill = NA))
          grid.circle(x=x,y=y,r= percent_mat[i, j]/2 * min(unit.c(w, h)),
                      gp = gpar(fill = col_fun(exp_mat[i, j]), col = NA))}

#Function to set dot color based on mean expression
col_fun = circlize::colorRamp2(c(min(exp_mat),(max(exp_mat)-min(exp_mat))/2, max(exp_mat)), viridis(100)[c(1,50,100)])

#Function to set legend size according to percentage expressed
layer_fun = function(j, i, x, y, w, h, fill){
          grid.rect(x = x, y = y, width = w, height = h, 
                    gp = gpar(col = NA, fill = NA))
          grid.circle(x=x,y=y,r= pindex(percent_mat, i, j)/1 * unit(1.5, "mm"),
                      gp = gpar(fill = col_fun(pindex(exp_mat, i, j)), col = NA))}

#Generate legend for percentage expressed
lgd_list = list(
    Legend( labels = c(0,0.25,0.5,0.75,1), title = "percentage expressed",
            graphics = list(
              function(x, y, w, h) grid.circle(x = x, y = y, r = 0 * unit(2, "mm"),
                                               gp = gpar(fill = "black")),
              function(x, y, w, h) grid.circle(x = x, y = y, r = 0.25 * unit(2, "mm"),
                                               gp = gpar(fill = "black")),
              function(x, y, w, h) grid.circle(x = x, y = y, r = 0.5 * unit(2, "mm"),
                                               gp = gpar(fill = "black")),
              function(x, y, w, h) grid.circle(x = x, y = y, r = 0.75 * unit(2, "mm"),
                                               gp = gpar(fill = "black")),
              function(x, y, w, h) grid.circle(x = x, y = y, r = 1 * unit(2, "mm"),
                                               gp = gpar(fill = "black")))
            ))

#Heatmap
hp <- Heatmap(exp_mat,
        heatmap_legend_param=list(title="expression"),
        column_title = "Dotplot - normalized mean expression and %expressed", 
        top_annotation = column_ha, 
        col=col_fun,
        rect_gp = gpar(type = "none"),
        clustering_distance_columns="pearson", #use euclidean, kendall or pearson
        clustering_method_columns="average", #use centroid, average or single
        layer_fun = layer_fun,
        row_labels=sapply(strsplit(rownames(exp_mat), "_"), "[[", 1),
        cell_fun = cell_fun,
        show_row_dend = F,
        column_dend_height = unit(1.5, "cm"),
        row_names_gp = gpar(fontsize = 10),
        column_names_gp = gpar(fontsize = 10),
        column_names_rot = 45,
        border = "black")

d1 <- draw( hp, annotation_legend_list = lgd_list)

setEPS()
postscript(file.path(path, "results/Manuscript/1_Figure/R/bubbleplot_scaled.eps"), height=7.2, width=14)
d1
dev.off()
```

Next we will visualize the tissue type in the UMAP space.

```{r visualizing-umap-by-phenograph-cluster, message=F, fig.width=17}
mat_red <- reducedDim(spe, "UMAP") 
tissue <- spe$tissue
mat_red_df <- as.data.frame(mat_red)
mat_red_df["tissue"] <- tissue
mat_red_df_rand <- mat_red_df[sample(1:nrow(mat_red_df)),]
  
plot <- ggplot(mat_red_df_rand, aes(x=V1, y=V2, colour=tissue)) +
    geom_point(alpha=0.5, show.legend=T) + 
    ggtitle("Tissue type") +
    scale_colour_manual(values = col_tissue) +
    theme(plot.title = element_text(size = 15), 
          axis.title = element_text(size=15), 
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(), 
          panel.background = element_blank(), 
          axis.line = element_line(colour = "black"),
          legend.text = element_text(size=15))

plot
```

Next we will visualize the clusters in the UMAP space.

```{r visualizing-umap-by-phenograph-cluster, message=F, fig.width=17}

mat_red_df <- as.data.frame(mat_red)
mat_red_df["celltype"] <- celltype
mat_red_df_rand <- mat_red_df[sample(1:nrow(mat_red_df)),]
  
plot <- ggplot(mat_red_df_rand, aes(x=V1, y=V2, colour=celltype)) +
    geom_point(show.legend=T) + 
    ggtitle("Phenograph cluster") +
    scale_colour_manual(values = col_celltype) +
    theme(plot.title = element_text(size = 15), 
          axis.title = element_text(size=15), 
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(), 
          panel.background = element_blank(), 
          axis.line = element_line(colour = "black"),
          legend.text = element_text(size=15))

plot
```

Now, we will plot all samples in the dim-red space. 

```{r visualizing-umap-by-sample, message=F, fig.width=20}

library(RColorBrewer)

rownames(mat_red) <- spe$sample
mat_red_df <- as.data.frame(mat_red)
mat_red_df["sample"] <- rownames(mat_red)
mat_red_df_rand <- mat_red_df[sample(1:nrow(mat_red_df)),]
  
plot <- ggplot(mat_red_df_rand, aes(x=V1, y=V2, colour=sample)) +
    geom_point(size=0.05, alpha=0.5, show.legend=F) + 
    ggtitle("Sample") +
    scale_colour_manual(values = col_sample) +
    theme(plot.title = element_text(size = 5), 
          axis.title = element_text(size=5), 
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(), 
          panel.background = element_blank(), 
          axis.line = element_line(colour = "black"),
          legend.text = element_text(size=5))

plot
```

Now, we will plot the marker separately in the dim-red space. 

```{r visualizing-umap-by-marker, message=F, fig.width=20}
#mat_tmp <- rbind(mat, intensity_mat_lq_wo_zeros, morph_mat_clipped)
mat_tmp <- morph_mat_clipped[c("area", "solidity"),]

for (i in 1: nrow(mat_tmp)){
  marker <- rownames(mat_tmp)[i]
  mat_red_df <- as.data.frame(mat_red)
  mat_red_df["marker"] <- mat_tmp[i,]
  mat_red_df_rand <- mat_red_df[sample(1:nrow(mat_red_df)),]
  #mat_red_df_rand <- mat_red_df[order(mat_red_df["marker"], decreasing=T),]
    
  plot <- ggplot(mat_red_df_rand, aes(x=V1, y=V2, colour=marker)) +
      geom_point(size=0.05, alpha=0.2, show.legend=T) + 
      ggtitle(marker) +
      scale_colour_viridis_c() +
      theme(plot.title = element_text(size = 5), 
            axis.title = element_text(size=5), 
            panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(), 
            panel.background = element_blank(), 
            axis.line = element_line(colour = "black"),
            legend.text = element_text(size=5))
  
  ggsave(filename=file.path(path, paste0("results/Manuscript/1_Figure/", marker , ".tiff")), width = 10, height = 5) #units = "in", device='tiff', dpi=500)
} 
```

Next, we will plot the tissue type separately in the dim-red space.

```{r visualizing-umap-by-tissuetype}

var <- "tissue"

rownames(mat_red) <- spe$sample_id
meta <- colData(spe)[match(rownames(mat_red), spe$sample_id),var]

samples <- unique(meta)

bin_col <- setNames(c(adjustcolor("gray90",alpha.f=1), "palegreen4"), c(FALSE,TRUE))

plot_list <- list()

for(i in 1: length(samples)){

  sample_binary <- meta == samples[i] 
  
  background <- as.data.frame(mat_red[!sample_binary,])
  foreground <- as.data.frame(mat_red[sample_binary,])
  plot_list[[i]] <- ggplot(background, aes(x=V1, y=V2) ) +
    geom_point(color="gray90", size=0.02) +
    geom_point(data=foreground, color="palegreen4", show.legend=F, size=0.02, alpha=0.25) +
    ggtitle(samples[i]) +
    theme(plot.title = element_text(size = 5), 
          axis.title = element_text(size=5), 
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(), 
          panel.background = element_blank(), 
          axis.line = element_line(colour = "black"))
}


plot_list[[1]]
plot_list[[2]]
```

Now, we will plot the clusters separately in the dim-red space. 

```{r visualizing-umap-by-marker, message=F, fig.width=20}

samples <- unique(celltype)

for(i in 1: length(samples)){

  sample_binary <- celltype == samples[i] 
  
  background <- as.data.frame(mat_red[!sample_binary,])
  foreground <- as.data.frame(mat_red[sample_binary,])
  plot <- ggplot(background, aes(x=V1, y=V2) ) +
    geom_point(color="gray90", size=0.02) +
    geom_point(data=foreground, color="palegreen4", show.legend=F, size=0.02, alpha=0.25) +
    ggtitle(samples[i]) +
    theme(plot.title = element_text(size = 5), 
          axis.title = element_text(size=5), 
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(), 
          panel.background = element_blank(), 
          axis.line = element_line(colour = "black"))
  
  ggsave(filename=file.path(path, paste0("results/Manuscript/1_Figure/", samples[i] , ".tiff")), width = 10, height = 5, units = "in", device='tiff', dpi=500)
}

```

Now, we will plot the clusters separately in the dim-red space per PT and BM. 

```{r visualizing-umap-by-marker, message=F, fig.width=20}

celltype <- spe$celltype
mat_red_df <- as.data.frame(mat_red)
mat_red_df["tissue"] <- spe$tissue

samples <- unique(celltype)

for(i in 1: length(samples)){

  sample_binary <- celltype == samples[i] 
  
  background <- as.data.frame(mat_red_df[!sample_binary,])
  foreground <- as.data.frame(mat_red_df[sample_binary,])
  plot <- ggplot(background, aes(x=V1, y=V2) ) +
    geom_point(color="gray90", size=0.02) +
    geom_point(data=foreground, aes(x=V1, y=V2, colour=tissue), show.legend=F, size=0.02, alpha=0.25) +
    scale_colour_manual(values = col_tissue) +
    ggtitle(samples[i]) +
    theme(plot.title = element_text(size = 5), 
          axis.title = element_text(size=5), 
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(), 
          panel.background = element_blank(), 
          axis.line = element_line(colour = "black"))
  
  ggsave(filename=file.path(path, paste0("results/Manuscript/1_Figure/", samples[i] , ".tiff")), width = 10, height = 5, units = "in", device='tiff', dpi=500)
}

```


Finally, we will visualize the data in the dimensionality reduced space and color the dots by the sample they belong to. 

```{r visualizing-umap-by-sample-single}

var <- "sample"
colData(spe) <- spe$sample
meta <- colData(spe)[match(rownames(mat_red), spe$sample),var]

samples <- unique(meta)

bin_col <- setNames(c(adjustcolor("gray90",alpha.f=1), "palegreen4"), c(FALSE,TRUE))

plot_list <- list()

for(i in 1: length(samples)){

  sample_binary <- meta == samples[i] 
  
  background <- as.data.frame(mat_red[!sample_binary,])
  foreground <- as.data.frame(mat_red[sample_binary,])
  plot_list[[i]] <- ggplot(background, aes(x=V1, y=V2) ) +
    geom_point(color="gray90", size=0.05) + 
    geom_point(data=foreground, color="palegreen4", show.legend=F, size=0.05, alpha=0.25) +
    ggtitle(samples[i]) +
    theme(plot.title = element_text(size = 5), 
          axis.title = element_text(size=5), 
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(), 
          panel.background = element_blank(), 
          axis.line = element_line(colour = "black"))
}


plot_sublists <- split(plot_list, ceiling(seq_along(plot_list) / 33)) 

save_tiff <- function(sublist, j){
  plot <- plot_grid(plotlist = sublist, nrow=11, ncol=3) 
  ggsave(filename=file.path(path, paste("/results/Manuscript/1_Figure/sample_", j, ".tiff", sep="")), width = 7, height = 20, units = "in", device='tiff', dpi=500)
}

mapply(save_tiff, plot_sublists, 1:length(plot_sublists))
```
Next, we will plot the number of DTCs detected by IF (diagnostics) and IMC.
```{r IF-vs-IMC, fig.width=12}
df <- distinct(data.frame(sample=spe$sample, IF_right=spe$infiltration_right, IF_left=spe$infiltration_left))
df$IMC_DTC_count <- unlist(lapply(df_test$sample, function(x){length(spe[,spe$sample==x & spe$metacluster=="tumor"]$sample_id)}))
df$IMC_cell_count <- unlist(lapply(df_test$sample, function(x){length(spe[,spe$sample==x]$sample_id)}))
df$IMC_percentage <- (df$IMC_DTC_count / df$IMC_cell_count)
write.csv(df, file.path(path, "results/Manuscript/1_Figure/tumor_infiltration_IF_vs_IMC.csv"))
```

Next, we will visualize the cluster per sample and the sample per cluster as described [here](https://r-graph-gallery.com/48-grouped-barplot-with-ggplot2.html).
```{r barplot-cluster-per-sample, fig.width=12}
 # library
library(ggplot2)
library(tidyverse)

celltype <- spe$celltype
col_celltype <- metadata(spe)$color_vectors$col_celltype
col_celltype<- col_celltype[order(as.integer(names(col_celltype)))]

colnames(mat) <- spe$sample

#for more informative sample names
#colnames(mat) <- paste0(spe$tissue, "_", "MYCN:", spe$MYCN_amp, "_time:", spe$timepoint, "_Prog:", spe$progression, "_crtl:", spe$control, "_", spe$fm_id)

data <- data.frame(sample=colnames(mat), celltype)
data$freq <- 1

data_wide <- data %>% pivot_wider(names_from = c("celltype"), values_from = "freq", values_fn = sum, values_fill = 0) 

#write.csv(data_wide, file.path(path, "results/Manuscript/1_Figure/celltype-per-sample.csv"))
data_long <- data_wide %>% pivot_longer(cols = !sample, names_to = "celltype", values_to = "count") 
data_long <- data_long[order(data_long$sample),]
data_long <- data_long[grepl("TU", data_long$sample),] #& (grepl("MYCN:1", data_long$sample)| grepl("MYCN:3", data_long$sample)),]

#data_long$celltype <- factor(data_long$celltype, levels=levels(celltype))
data_long$celltype <- factor(data_long$celltype)
 
# Stacked + percent
ggplot(data_long, aes(fill=celltype, y=count, x=sample)) + 
  ggtitle("Clusters per PT sample") +
  geom_bar(position="fill", stat="identity") + 
  scale_fill_manual(values = col_celltype) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 
```


```{r barplot-samples-per-cluster, fig.width=20}
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector_433 = colors()[grep('gr(a|e)y', grDevices::colors(), invert = T)]
col_sample = setNames(sample(col_vector_433, length(unique(data$sample))), unique(data$sample))

data_wide <- data %>% pivot_wider(names_from = c("sample"), values_from = "freq", values_fn = sum, values_fill = 0)

#write.csv(data_wide, file.path(path, "results/Manuscript/1_Figure/samples-per-celltype.csv"))

data_long <- data_wide %>% pivot_longer(cols = !celltype, names_to = "sample", values_to = "count") 

 
# Stacked + percent
ggplot(data_long, aes(fill=sample, y=count, x=celltype)) + 
  ggtitle("Samples per celltype") +
  geom_bar(position="fill", stat="identity", show.legend = F) + 
  scale_fill_manual(values = col_sample) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 
```

We will now visualize the tissues per cluster. 

```{r barplot-tissues-per-cluster, fig.width=20}

colnames(mat) <- spe$sample
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector_433 = colors()[grep('gr(a|e)y', grDevices::colors(), invert = T)]

data <- data.frame(tissue=gsub(".*_*_", "", colnames(mat)), celltype)
data[data$tissue%in%"TU",]$tissue <- "PT"
data$freq <- 1
data_wide <- data %>% pivot_wider(names_from = c("tissue"), values_from = "freq", values_fn = sum, values_fill = 0)

write.csv(data_wide, file.path(path, "/results/Manuscript/1_Figure/tissues-per-celltype.csv"))

data_long <- data_wide %>% pivot_longer(cols = !celltype, names_to = "tissue", values_to = "count") 

 
# Stacked + percent
ggplot(data_long, aes(fill=tissue, y=count, x=celltype)) + 
  ggtitle("Samples per celltype") +
  geom_bar(position="fill", stat="identity", show.legend = T) + 
  scale_fill_manual(values = col_tissue) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 
```

We will now visualize the clusters in a few samples to see if they make sense.

```{r visualize-clusters-in-image, message=FALSE, fig.width=15}
library(cytomapper)

tissue <- "TU"

#files <- list.files(file.path(path, "results/in_house/colored_masks/masks_temp", tissue))
files <- list.files(file.path(path, "masks_temp", tissue))

for(i in 1: length(files)){
#for(i in 1: 1){
  
  #masks <- loadImages(file.path(path, "masks_temp", tissue, files[i]), as.is = TRUE)
  masks <- loadImages(file.path(path, "masks_temp", tissue, files[i]), as.is = TRUE)
  
  # tmp <- sub('.*clust_', '', files[i])
  # names(masks) <- sub("\\.tif.*", "", tmp)
  # c <- sub("\\_clust.*", "", files[i])
  
  c <- "early SYM-like TC"
  mcols(masks) <- DataFrame(sample_id = names(masks))
  
  plotCells(masks,
            object = spe,
            cell_id = "ObjectNumber", img_id = "sample_id",
            colour_by = "celltype",
            colour = list(celltype = metadata(spe)$color_vectors$col_celltype_2),
            display = "single",
            save_plot = list(filename = file.path(path, "tmp", tissue, paste0("yellow_", files[i], "f"))),
            #save_plot = list(filename = file.path(path, "tmp", tissue, paste0("all_", files[i], "f"))),
            legend = NULL)


  col_celltype_bin <- setNames(c("red", "white"), c('yes', 'no'))
  spe$binary <- 'no'
  spe[,spe$celltype == c]$binary <- 'yes'


  # plotCells(masks,
  #           object = spe,
  #           cell_id = "ObjectNumber", img_id = "sample_id",
  #           colour_by = "binary",
  #           colour = list(binary = col_celltype_bin),
  #           display="single",
  #           save_plot = list(filename = file.path(path, "tmp", tissue, paste0(files[i], "f"))),
  #           #save_plot = list(filename = file.path(path, "tmp", tissue, paste0(files[i], "f"))),
  #           legend=NULL)

}

```

Next, we will look at cluster abundance per sample and visualize it together with the clinical data in a heatmap. 

```{r cluster-abundance-per-sample, fig.width=15, fig.height=12}
spe <- spe_all[,spe_all$tissue=="PT"]
data <- data.frame(sample=spe$sample, celltype=spe$celltype)
data$freq <- 1
data_wide <- as.data.frame(data %>% pivot_wider(names_from = c("celltype"), values_from = "freq", values_fn = sum, values_fill = 0))
rownames(data_wide) <- data_wide$sample
data_wide <- data_wide[,!colnames(data_wide) %in% c("sample")]
proportions <- apply(data_wide, 1, prop.table)

#scaled_prop <- t(apply(proportions, 1, RESCALE, to=0:1))
scaled_prop <- proportions

mean_imc_score <- aggregate(spe$imc_score, list(spe$sample), mean)
mean_IMC_score <- mean_imc_score[match(colnames(scaled_prop), mean_imc_score$Group.1),]$x
col_mean_IMC_score = colorRamp2(c(min(mean_IMC_score),min(mean_IMC_score)+(max(mean_IMC_score)-min(mean_IMC_score))/2, max(mean_IMC_score)), c("blue", "white", "red"))

coldata <- data.frame(colData(spe)[,c("sample", "stainer", "control", "study_id", "timepoint", "stage", "progression", "efs_dur", "MYCN_amp", "ALK_amp", "ALK_mut", "ATRX_del", "TERT_rear", "X1p_loss", "X1q_gain", "X11q_loss", "X17q_gain")]) %>%
    group_by(sample, stainer, control, study_id, timepoint, stage, progression, efs_dur, MYCN_amp, ALK_amp, ALK_mut, ATRX_del, TERT_rear, X1p_loss, X1q_gain, X11q_loss, X17q_gain) %>% summarise(n=n(), .groups="drop")
coldata <- coldata[match(colnames(scaled_prop), coldata$sample),]

column_ha = HeatmapAnnotation(imc_score=mean_IMC_score,
                              control = coldata$control,
                              patient = coldata$study_id,
                              timepoint = coldata$timepoint,
                              stage = coldata$stage,
                              progression = coldata$progression, 
                              MYCN_amp = coldata$MYCN_amp, 
                              ALK_amp = coldata$ALK_amp, 
                              ALK_mut = coldata$ALK_mut, 
                              ATRX_del = coldata$ATRX_del, 
                              TERT_rear = coldata$TERT_rear, 
                              X1p_loss = coldata$X1p_loss, 
                              X1q_gain = coldata$X1q_gain, 
                              X11q_loss = coldata$X11q_loss, 
                              X17q_gain = coldata$X17q_gain, 
                              ncells=anno_barplot(coldata$n, height = unit(3, "cm")),
                              col=list(imc_score=col_mean_IMC_score, 
                                      control = metadata(spe)$color_vectors$col_control,
                                      patient = metadata(spe)$color_vectors$col_study_id,
                                      timepoint = metadata(spe)$color_vectors$col_timepoint,
                                      stage = metadata(spe)$color_vectors$col_stages,
                                      progression = metadata(spe)$color_vectors$col_progress, 
                                      MYCN_amp = metadata(spe)$color_vectors$col_aberrations, 
                                      ALK_amp = metadata(spe)$color_vectors$col_aberrations, 
                                      ALK_mut = metadata(spe)$color_vectors$col_aberrations, 
                                      ATRX_del = metadata(spe)$color_vectors$col_aberrations, 
                                      TERT_rear = metadata(spe)$color_vectors$col_aberrations, 
                                      X1p_loss = metadata(spe)$color_vectors$col_aberrations, 
                                      X1q_gain = metadata(spe)$color_vectors$col_aberrations, 
                                      X11q_loss = metadata(spe)$color_vectors$col_aberrations, 
                                      X17q_gain = metadata(spe)$color_vectors$col_aberrations), 
                              show_legend=T)

Heatmap(scaled_prop,
        column_labels=paste0(coldata$study_id, "_", coldata$sample),
        column_names_max_height = unit(3, "cm"),
        clustering_method_columns = "complete",
        clustering_distance_columns = "euclidean",
        top_annotation = column_ha,
        column_dend_height = unit(5, "cm"),
        cluster_columns=T,
        cluster_rows=T)
```

We will show the frequency of clusters in control and infiltrated samples in stacked barplots. 
```{r differential-cluster-abundance, message=FALSE, fig.height=20, fig.width=20}
data <- data.frame(celltype=spe$celltype, control=spe$control)
data$freq <- 1
data_wide <- as.data.frame(data %>% pivot_wider(names_from = c("control"), values_from = "freq", values_fn = sum, values_fill = 0))
rownames(data_wide) <- data_wide$celltype
data_wide <- data_wide[,!colnames(data_wide) %in% c("celltype")]
control_proportions <- t(apply(data_wide, 1, prop.table))
```

We will show the tissue type in stacked barplots. 
```{r differential-cluster-abundance, message=FALSE, fig.height=20, fig.width=20}
data <- data.frame(celltype=spe$celltype, tissue=spe$tissue)
data$freq <- 1
data_wide <- as.data.frame(data %>% pivot_wider(names_from = c("tissue"), values_from = "freq", values_fn = sum, values_fill = 0))
rownames(data_wide) <- data_wide$celltype
data_wide <- data_wide[,!colnames(data_wide) %in% c("celltype")]
tissue_proportions <- t(apply(data_wide, 1, prop.table))
```

Next, we will calculate the means per cluster.

```{r mean-per-cluster-with-metadata, message=FALSE, fig.height=20, fig.width=40}
library(scuttle)
library(bruceR)

set.seed(20230135)

matrix_hq <- mat_hq
matrix_lq <- mat_lq
morph_matrix <- mat_morph
clust <- celltype

#Mean of high quality markers per cluster
mean <- aggregate(t(matrix_hq), list(clust), mean)
mean_t <- t(mean[,2:ncol(mean)])
mean_t <- t(apply(mean_t, 1, RESCALE, to=0:1))
colnames(mean_t) <- mean[,"Group.1"]

#Mean of morphological features per cluster
mean_morph <- aggregate(t(morph_matrix), list(clust), mean)
area_mean = mean_morph[,"area"]
solidity_mean = mean_morph[,"solidity"]
aspect_ratio_mean = mean_morph[,"aspect_ratio"]

#Mean of low quality markers per cluster
mean_lq <- aggregate(t(matrix_lq), list(clust), mean)
clPARP_mean = mean_lq[,"clPARP_Yb176_mean"]

#Mean of low metadata per cluster
mean_metaclust <-aggregate(spe$metacluster, list(clust), max)
mean_metacluster <- mean_metaclust$x

col_mean_area = colorRamp2(c(min(area_mean),min(area_mean)+(max(area_mean)-min(area_mean))/2, max(area_mean)), c("blue", "white", "red"))
col_mean_solidity = colorRamp2(c(min(solidity_mean),min(solidity_mean)+(max(solidity_mean)-min(solidity_mean))/2, max(solidity_mean)), c("blue", "white", "red"))
col_mean_aspect_ratio = colorRamp2(c(min(aspect_ratio_mean),min(aspect_ratio_mean)+(max(aspect_ratio_mean)-min(aspect_ratio_mean))/2, max(aspect_ratio_mean)), c("blue", "white", "red"))
col_mean_clPARP = colorRamp2(c(min(clPARP_mean),min(clPARP_mean)+(max(clPARP_mean)-min(clPARP_mean))/2, max(clPARP_mean)), c("blue", "white", "red"))
```

Next, we want to look at differential cluster abundance between MNA and nMNA groups.
```{r differential-cluster-abundance, message=FALSE, fig.height=20, fig.width=20}
pos_samples <- coldata[which(coldata$MYCN_amp==1 | coldata$MYCN_amp==3),"sample"]$sample
neg_samples <- coldata[which(coldata$MYCN_amp==0),"sample"]$sample

MNA <- proportions[match(mean[,"Group.1"], rownames(proportions)), pos_samples]
nMNA <- proportions[match(mean[,"Group.1"], rownames(proportions)), neg_samples]

rg = range(c(MNA, nMNA))
rg[1] = rg[1] - (rg[2] - rg[1])* 0.02
rg[2] = rg[2] + (rg[2] - rg[1])* 0.02
anno_mna_boxplot = function(index) {
    nr = length(index)
    pushViewport(viewport(xscale = rg, yscale = c(0.5, nr + 0.5)))
    for(i in seq_along(index)) {
        grid.rect(y = nr-i+1, height = 1, default.units = "native")
        grid.boxplot(MNA[ index[i], ], pos = nr-i+1 + 0.2, box_width = 0.3, 
            gp = gpar(fill = "red"), direction = "horizontal")
        grid.boxplot(nMNA[ index[i], ], pos = nr-i+1 - 0.2, box_width = 0.3, 
            gp = gpar(fill = "green"), direction = "horizontal")
      }
    
    grid.xaxis()
    popViewport()
}
```

Then we will look at the mean marker expression per cluster and plot metadata in the annotation bars.
```{r mean-per-cluster-heatmap-with-metadata, message=FALSE, fig.height=14, fig.width=20}

row_ha = rowAnnotation(metacluster=mean_metacluster,
                              celltype = mean[,"Group.1"], 
                              area=area_mean,
                              solidity = solidity_mean,
                              aspect_ratio = aspect_ratio_mean,
                              clPARP = clPARP_mean,
                              ncells=anno_barplot(as.data.frame(table(clust))[,"Freq"], height = unit(5, "cm")),
                              mna=anno_mna_boxplot, width = unit(18, "cm"), 
                              control = anno_barplot(control_proportions[match(mean[,"Group.1"], rownames(control_proportions)),], gp = gpar(fill = unname(metadata(spe)$color_vectors$col_control[colnames(control_proportions)])), bar_width = 1, height = unit(6, "cm")),
                              tissue = anno_barplot(tissue_proportions[match(mean[,"Group.1"], rownames(tissue_proportions)),], gp = gpar(fill = unname(metadata(spe)$color_vectors$col_tissue[colnames(tissue_proportions)])), beside = TRUE, attach = TRUE),
                              col=list(metacluster=metadata(spe)$color_vectors$col_metacluster,
                                       celltype=metadata(spe)$color_vectors$col_celltype, 
                                       area=col_mean_area,
                                       solidity=col_mean_solidity,
                                       aspect_ratio=col_mean_aspect_ratio,
                                       clPARP = col_mean_clPARP
                                       ), 
                              show_legend=T)


h1 <- Heatmap(t(mean_t), 
        column_title = "Heatmap showing mean marker expression per cluster-scaled",
        name = "marker_expression", 
        column_labels=sapply(strsplit(rownames(mean_t), "_"), "[[", 1),
        left_annotation = row_ha, 
        cluster_columns=T,
        clustering_distance_rows="pearson", #use euclidian, kendal or pearson
        clustering_method_rows="single",
        cluster_rows=T,
        show_column_dend = T,
        show_column_names=T,
        show_row_names=T,
        col=viridis(100),
        row_dend_width = unit(5, "cm"),
        column_dend_height = unit(2, "cm"),
        column_names_gp = gpar(fontsize = 8),
        )

h2 <- Heatmap(scaled_prop[match(colnames(mean_t), rownames(scaled_prop)),],
        column_title = "Heatmap showing cluster abundance per sample",
        name = "cluster_abundance", 
        column_labels=paste0(coldata$study_id, "_", coldata$sample),
        column_names_max_height = unit(12, "cm"),
        top_annotation = column_ha,
        column_dend_height = unit(10, "cm"),
        cluster_columns=T,
        cluster_rows=F,
        column_names_gp = gpar(fontsize = 8))

lgd1 = Legend(labels = c("pos/het", "neg"), title = "MYCN_amp",
    legend_gp = gpar(fill = c("red", "green")))

lgd2 = Legend(labels = c("BM", "PT"), title = "tissue",
    legend_gp = gpar(fill = unname(metadata(spe)$color_vectors$col_tissue[colnames(tissue_proportions)])))

draw(h1+h2, heatmap_legend_list = list(lgd1,lgd2))

```

```{r mean-per-cluster-heatmap-with-metadata, message=FALSE, fig.height=6, fig.width=8}
clust <- spe$celltype
clust_pt <- spe[,spe$tissue=="PT"]$celltype
clust_bm <- spe[,spe$tissue=="BM"]$celltype


col_ha = columnAnnotation(#metacluster=mean_metacluster,
                              celltype = mean[mean[,"Group.1"]%in%unique(spe[,spe$metacluster=="tumor"]$celltype),"Group.1"], 
                              ncells_overall=anno_barplot(as.data.frame(table(clust))[,"Freq"], add_numbers=TRUE, height = unit(1.8, "cm")),
                              ncells_PT=anno_barplot(as.data.frame(table(clust_pt))[,"Freq"], add_numbers=TRUE, height = unit(1.8, "cm"), gp = gpar(fill = "brown")),
                              ncells_bm=anno_barplot(as.data.frame(table(clust_bm))[,"Freq"], add_numbers=TRUE, height = unit(1.8, "cm"), gp = gpar(fill = "aquamarine4")),
                              tissue = anno_barplot(tissue_proportions[match(mean[mean[,"Group.1"]%in%unique(spe[,spe$metacluster=="tumor"]$celltype),"Group.1"], rownames(tissue_proportions)),], gp = gpar(fill = unname(metadata(spe)$color_vectors$col_tissue[colnames(tissue_proportions)])), beside = TRUE, attach = TRUE),
                              col=list(metacluster=metadata(spe)$color_vectors$col_metacluster,
                                       celltype=metadata(spe)$color_vectors$col_celltype
                                       ), 
                              show_legend=T)


h1 <- Heatmap(mean_t, 
        #title = "Heatmap showing mean marker expression per cluster-scaled",
        name = "marker_expression", 
        row_labels=sapply(strsplit(rownames(mean_t), "_"), "[[", 1),
        top_annotation = col_ha, 
        cluster_rows=T,
        clustering_distance_columns="pearson", #use euclidian, kendal or pearson
        clustering_method_columns="single",
        cluster_columns=F,
        show_row_dend = F,
        show_column_dend = F,
        show_row_names=T,
        show_column_names=T,
        col=viridis(100),
        #row_dend_width = unit(5, "cm"),
        #column_dend_height = unit(2, "cm"),
        row_names_gp = gpar(fontsize = 8),
        column_names_rot = 45,
        column_names_gp = grid::gpar(fontsize = 6)
        )

lgd = Legend(labels = c("BM", "PT"), title = "tissue",
    legend_gp = gpar(fill = unname(metadata(spe)$color_vectors$col_tissue[colnames(tissue_proportions)])))

draw(h1, heatmap_legend_list = list(lgd))

```
