---
title: "Read in scRNAseq data"
output: html_document
date: '2023-07-06'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Read in the single-cell data {#read-data}

Here we will read in single-cell RNAseq data from a single-cell adrenal gland atlas by 
[Jansky et al.](https://adrenal.kitz-heidelberg.de/developmental_programs_NB_viz/), and
our imaging-based single-cell data

```{r read-data, message=FALSE}
library(Seurat)
library(SingleCellExperiment)
library(dplyr)
library(tidyr)
library(patchwork)

path <- "/mnt/Multimodal_Imaging_Daria/PIPELINE/NEWEST_RESULTS"

reference <- readRDS(file.path(path,"public_datasets/Triana/Healthy.rds")) #Healthy bone marrow 97plex
reference2 <- readRDS(file.path(path,"public_datasets/Triana/200AB_projected.rds")) #Healthy bone marrow and PBMCs - 197plex

rna <- reference2
DefaultAssay(rna) <- "AB"
rna@active.ident <- rna@meta.data$ct

protein <- readRDS(file.path(path,"results/spe_clustered.rds"))
protein <- protein[rowData(protein)$use_channel,]
col_celltype <- metadata(protein)$color_vectors$col_celltype
```

# Set common features 
Next, we will set common features between the reference and query dataset. Therefore,
we have to rename the features in the imaging dataset to the names of the corresponding
genes. 

```{r set common features, message=FALSE}

protein2gene <- read.csv(file.path(path,"public_datasets/Triana/protein2gene.csv"), sep=";")

code <- setNames(protein2gene$prot_triana, protein2gene$protein)
rownames(protein) <- recode(rownames(protein), !!!code)
rowData(protein)$name <- rownames(protein)
rowData(protein)$channel <- rownames(protein)

protein <- as.Seurat(protein, counts = "counts", data = "exprs")
Idents(protein) <- "celltype"

common_features <- intersect(row.names(rna), row.names(protein))
```


We will then visualize the two datasets by UMAP.

```{r umap-rna, fig.width=18, fig.height=7, message=FALSE}
library(RColorBrewer)
library(paletteer)
set.seed(230713)

col_vector_433 = colors()[grep('gr(a|e)y', grDevices::colors(), invert = T)]
col_rna_clusters = setNames(sample(col_vector_433, length(unique(rna@active.ident))), unique(rna@active.ident))

#DimPlot(rna, reduction="ProjectedUMAP", cols=col_rna_clusters)
DimPlot(rna, reduction="umap", cols=col_rna_clusters)

```


```{r umap-protein, message=FALSE, fig.width=20, fig.height=15}
set.seed(230713)
col_protein_clusters = setNames(sample(col_vector_433, length(unique(protein@active.ident))), unique(protein@active.ident))
DimPlot(protein, reduction="UMAP", cols=col_protein_clusters)
```

```{r mean-per-cluster-rna, fig.height=15, message=FALSE}
library(bruceR)
library(ComplexHeatmap)
library(viridis)

rna_clusters <- rna@active.ident
rna_data <- GetAssayData(rna, assay = "AB", slot = "data")[common_features, ]
rna_mean <- aggregate(t(rna_data), list(rna_clusters), mean)

rna_mean_t <- t(rna_mean[,2:ncol(rna_mean)])
colnames(rna_mean_t) <- rna_mean[,"Group.1"]
rna_mean_scaled <- t(apply(rna_mean_t, 1, RESCALE, to=0:1))

column_ha = HeatmapAnnotation(ncells=anno_barplot(as.data.frame(table(rna_clusters))[,"Freq"], height = unit(3, "cm")),
                              show_legend=F)

Heatmap(rna_mean_scaled, 
        column_title = "Mean per cluster (RNA data)",
        top_annotation = column_ha, 
        cluster_columns=T,
        show_column_dend = T,
        show_column_names=T,
        show_row_names=T,
        col=viridis(100)
        )

```

```{r mean-per-cluster-protein, fig.height=15, message=FALSE}

protein_clusters <- protein@active.ident
protein_data <- GetAssayData(protein, assay = "originalexp", slot = "data")[common_features, ]
protein_mean <- aggregate(t(protein_data), list(protein_clusters), mean)

protein_mean_t <- t(protein_mean[,2:ncol(protein_mean)])
colnames(protein_mean_t) <- protein_mean[,"Group.1"]
protein_mean_scaled <- t(apply(protein_mean_t, 1, RESCALE, to=0:1))

column_ha = HeatmapAnnotation(ncells=anno_barplot(as.data.frame(table(protein_clusters))[,"Freq"], height = unit(3, "cm")),
                              show_legend=F)

Heatmap(protein_mean_scaled, 
        column_title = "Mean per cluster (protein data)",
        top_annotation = column_ha, 
        cluster_columns=T,
        show_column_dend = T,
        show_column_names=T,
        show_row_names=T,
        col=viridis(100)
        )

```
For further analysis we will exclude sticky cells and tumor cells.
```{r data-curation, message=FALSE}
protein <- subset(x=protein, idents = c("53_sticky cells"), invert=TRUE)

protein <- protein[,protein@meta.data$metacluster!="tumor"]

```

We will next plot the two datasets together in a heatmap to look at similarity.
```{r mean-per-cluster-combined, fig.height=15, message=FALSE}

clusters <- c(rna_clusters, protein_clusters)
cell_count <- c(as.data.frame(table(rna_clusters))[,"Freq"], as.data.frame(table(protein_clusters))[,"Freq"])
modality <- c(rep(c("RNA"), ncol(rna_mean_scaled)), rep(c("protein"), ncol(protein_mean_scaled)))
mean_scaled <- cbind(rna_mean_scaled, protein_mean_scaled)

column_ha = HeatmapAnnotation(modality=modality,
                              ncells=anno_barplot(cell_count, height = unit(3, "cm")),
                              show_legend=F)

Heatmap(mean_scaled, 
        column_title = "Mean per cluster (RNA & protein)",
        top_annotation = column_ha, 
        cluster_columns=T,
        show_column_dend = T,
        show_column_names=T,
        show_row_names=T,
        col=viridis(100)
        )


```

We will then look at cluster similarity in the PCA plot.
```{r pca-plot, fig.height=20, fig.width=20, message=FALSE}
library("FactoMineR")
library("factoextra")

data <- as.data.frame(mean_scaled)

data <- t(data)
data.pca <- FactoMineR::PCA(data, graph=F)
data$modality <- as.factor(modality)

fviz_pca_ind(data.pca, repel=T, habillage=data$modality)
```
