---
title: "Read in scRNAseq data"
output: html_document
date: '2023-07-06'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Read in the single-cell data {#read-data}

Here we will read in single-cell RNAseq data from a single-cell adrenal gland atlas by 
[Jansky et al.](https://adrenal.kitz-heidelberg.de/developmental_programs_NB_viz/), and
our imaging-based single-cell data

```{r read-data, message=FALSE}
library(Seurat)
library(SingleCellExperiment)
library(dplyr)
library(tidyr)

path <- "/mnt/Multimodal_Imaging_Daria/PIPELINE/NEWEST_RESULTS"

rna <- readRDS(file.path(path,"public_datasets/Jansky/adrenal_gland_Seurat.RDS"))
DefaultAssay(rna) <- "RNA"

protein <- readRDS(file.path(path,"results/spe_clustered.rds"))
protein <- protein[rowData(protein)$use_channel,]
col_celltype <- metadata(protein)$color_vectors$col_celltype
```

# Set common features 
Next, we will set common features between the reference and query dataset. Therefore,
we have to rename the features in the imaging dataset to the names of the corresponding
genes. 

```{r set common features, message=FALSE}

protein2gene <- read.csv(file.path(path,"public_datasets/protein2gene.csv"), sep=";")

code <- setNames(protein2gene$gene, protein2gene$protein)
rownames(protein) <- recode(rownames(protein), !!!code)
rowData(protein)$name <- rownames(protein)
rowData(protein)$channel <- rownames(protein)

protein <- as.Seurat(protein, counts = "counts", data = "exprs")
Idents(protein) <- "celltype"

common_features <- intersect(row.names(rna), row.names(protein))

# filtering cells without expression in common genes
indx <- colSums(rna[common_features,])!=0
rna <- rna[,indx]

indx <- colSums(protein[common_features,])!=0
protein <- protein[,indx]
```

# Quality control on public dataset

We will first look at cluster similarity to evaulate which clusters are similar. 
```{r mean-per-cluster-protein, fig.height=15, message=FALSE}
library(scuttle)
library(bruceR)
library(ComplexHeatmap)
library(viridis)

protein_clusters <- protein@active.ident
protein_data <- GetAssayData(protein, assay = "originalexp", slot = "data")[common_features, ]
protein_mean <- aggregate(t(protein_data), list(protein_clusters), mean)

protein_mean_t <- t(protein_mean[,2:ncol(protein_mean)])
colnames(protein_mean_t) <- protein_mean[,"Group.1"]
protein_mean_scaled <- t(apply(protein_mean_t, 1, RESCALE, to=0:1))

column_ha = HeatmapAnnotation(ncells=anno_barplot(as.data.frame(table(protein_clusters))[,"Freq"], height = unit(3, "cm")),
                              show_legend=F)

Heatmap(protein_mean_scaled, 
        column_title = "Mean per cluster (protein data)",
        top_annotation = column_ha, 
        cluster_columns=T,
        show_column_dend = T,
        show_column_names=T,
        show_row_names=T,
        col=viridis(100)
        )

```

```{r mean-per-cluster-rna, fig.height=15, message=FALSE}

rna_clusters <- rna@active.ident
rna_data <- GetAssayData(rna, assay = "RNA", slot = "data")[common_features, ]
rna_mean <- aggregate(t(rna_data), list(rna_clusters), mean)

rna_mean_t <- t(rna_mean[,2:ncol(rna_mean)])
colnames(rna_mean_t) <- rna_mean[,"Group.1"]
rna_mean_scaled <- t(apply(rna_mean_t, 1, RESCALE, to=0:1))

column_ha = HeatmapAnnotation(ncells=anno_barplot(as.data.frame(table(rna_clusters))[,"Freq"], height = unit(3, "cm")),
                              show_legend=F)

Heatmap(rna_mean_scaled, 
        column_title = "Mean per cluster (RNA data)",
        top_annotation = column_ha, 
        cluster_columns=T,
        show_column_dend = T,
        show_column_names=T,
        show_row_names=T,
        col=viridis(100)
        )

```

We will next plot the two together to look at similarity.
```{r mean-per-cluster-combined, fig.height=15, message=FALSE}

clusters <- c(rna_clusters, protein_clusters)
cell_count <- c(as.data.frame(table(rna_clusters))[,"Freq"], as.data.frame(table(protein_clusters))[,"Freq"])
modality <- c(rep(c("RNA"), ncol(rna_mean_scaled)), rep(c("protein"), ncol(protein_mean_scaled)))
mean_scaled <- cbind(rna_mean_scaled, protein_mean_scaled)

column_ha = HeatmapAnnotation(modality=modality,
                              ncells=anno_barplot(cell_count, height = unit(3, "cm")),
                              show_legend=F)

Heatmap(mean_scaled, 
        column_title = "Mean per cluster (RNA & protein)",
        top_annotation = column_ha, 
        cluster_columns=T,
        show_column_dend = T,
        show_column_names=T,
        show_row_names=T,
        col=viridis(100)
        )


```
We can see that immune cells are most similar to sticky cells and that CD8A does not correlate on RNA (expressed in tumor cells) and protein level. We will hence exclude the two. 

```{r mean-per-cluster-combined, fig.height=15, message=FALSE}

clusters_2 <- c(rna_clusters, protein_clusters[protein_clusters!="53_sticky cells"])
cell_count_2 <- as.data.frame(table(droplevels(clusters_2)))[,"Freq"]
modality_2 <- c(rep(c("RNA"), ncol(rna_mean_scaled)), rep(c("protein"), ncol(protein_mean_scaled[,colnames(protein_mean_scaled)!="53_sticky cells"])))
mean_scaled_2 <- cbind(rna_mean_scaled, protein_mean_scaled[,colnames(protein_mean_scaled)!="53_sticky cells"])

column_ha = HeatmapAnnotation(modality=modality_2,
                              ncells=anno_barplot(cell_count_2, height = unit(3, "cm")),
                              show_legend=F)

Heatmap(mean_scaled_2, 
        column_title = "Mean per cluster (RNA & protein)",
        top_annotation = column_ha, 
        cluster_columns=T,
        show_column_dend = T,
        show_column_names=T,
        show_row_names=T,
        col=viridis(100)
        )


```

We will then look at cluster similarity in the PCA plot.
```{r pca-plot, fig.height=20, fig.width=20, message=FALSE}
library("FactoMineR")
library("factoextra")

data <- as.data.frame(mean_scaled_2)

data <- t(data)
data.pca <- FactoMineR::PCA(data, graph=F)
data$modality <- as.factor(modality_2)

fviz_pca_ind(data.pca, repel=T, habillage=data$modality)
```


We will then visualize the two datasets by UMAP.

```{r umap-rna, message=FALSE}

DimPlot(rna, reduction="umap")
```


```{r umap-protein, message=FALSE, fig.width=20, fig.height=15}

DimPlot(protein, reduction="UMAP")
```

For further analysis we will exclude CD8A and the sticky cells.
```{r data-curation, message=FALSE}
protein <- subset(x=protein, idents = c("53_sticky cells"), invert=TRUE)

common_features <- common_features[common_features !="CD8A"]

indx <- colSums(rna[common_features,])!=0
rna <- rna[,indx]

indx <- colSums(protein[common_features,])!=0
protein <- protein[,indx]
```

# Batch correction

Next, we will try to integrate the two datasets with fastMNN and then calculate UMAP. 
```{r batch-correction, message=FALSE}
library(batchelor)
library(BiocParallel)
library(scater)

set.seed(220228)

#Integration of RNA and protein data
rna_mat <- GetAssayData(rna, assay = "RNA", slot = "data")[common_features, ]
protein_mat <- GetAssayData(protein, assay = "originalexp", slot = "data")[common_features, ]
mat <- cbind(rna_mat, protein_mat)
batch_id <- c(rep("rna", ncol(rna_mat)), rep("protein", ncol(protein_mat)))
clusters <- c(rna@active.ident, protein@active.ident)
out <- fastMNN(mat, batch = batch_id,
               auto.merge = F,
               BPPARAM = MulticoreParam())

mat_integrated <- reducedDim(out, "corrected")

saveRDS(mat_integrated, file.path(path,"public_datasets/Jansky/results/mat_integrated.rds"))
```

We will calculate UMAP on the integrated data. 

```{r dimensionality reduction, message=FALSE}

library(RColorBrewer)

#delete later - just doing this now to allow for faster parallelization
saveRDS(protein, file.path(path,"public_datasets/Jansky/results/protein.rds"))
saveRDS(rna, file.path(path,"public_datasets/Jansky/results/rna.rds"))
rm(protein)
rm(rna)

#UMAP and visualization
mat_t <- t(mat_integrated)
mat_red <- calculateUMAP(mat_t, BPPARAM = MulticoreParam())
saveRDS(mat_red, file.path(path,"public_datasets/Jansky/results/mat_red.rds"))

mat_red_df <- as.data.frame(mat_red)
mat_red_df$clusters <- clusters
mat_red_df$modality <- batch_id

saveRDS(mat_red_df, file.path(path,"public_datasets/Jansky/results/mat_red.rds"))
```

We will then visualize the integrated data in the UMAP space for RNA and protein separately. 

```{r batch-correction and dimensionality reduction, message=FALSE}
mod <- "rna"
binary <- mat_red_df$modality == mod
background <- as.data.frame(mat_red_df[!binary,])
foreground <- as.data.frame(mat_red_df[binary,])

set.seed(221228)

qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector_74 = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
col_vector_433 = colors()[grep('gr(a|e)y', grDevices::colors(), invert = T)]

col_clust = setNames(sample(col_vector_433, length(unique(mat_red_df$clusters))), unique(mat_red_df$clusters))
  
plot <- ggplot(background, aes(x=V1, y=V2)) +
    geom_point(color="gray90", size=0.02) +
    geom_point(data=foreground, aes(x=V1, y=V2, colour=clusters), show.legend=F, size=0.02, alpha=0.25) +
    scale_colour_manual(values = col_clusters) +
    ggtitle(paste0(mod, " clusters")) +
    theme(plot.title = element_text(size = 5), 
          axis.title = element_text(size=5), 
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(), 
          panel.background = element_blank(), 
          axis.line = element_line(colour = "black"))

plot
```

Next we will look at cluster similarity of the integrated data. 

```{r batch-correction and dimensionality reduction, message=FALSE}
#PCA to show cluster similarity
mean <- aggregate(t(mat_integrated), list(clusters), mean)

mean_t <- t(mean[,2:ncol(mean)])
colnames(mean_t) <- mean[,"Group.1"]
mean_scaled <- t(apply(mean_t, 1, RESCALE, to=0:1))

data <- as.data.frame(mean_scaled)
data <- t(data)
data.pca <- FactoMineR::PCA(data, graph=F)

modality <- rep("rna", ncol(data))
modality["_"%in%colnames(data)]<-"protein"
data$modality <- as.factor(modality)

fviz_pca_ind(data.pca, repel=T, habillage=data$modality)
```
## Multi-modal mapping

Here, we will first subset the datasets to a smaller size. 

```{r batch-correction and dimensionality reduction, message=FALSE}
#RNA
count <- min(table(rna_clusters))
rand_cells <- c()
for (i in 1: length(unique(rna_clusters))){
  idx <- which(rna_clusters == unique(rna_clusters)[i])
  rand_idx <- sample(idx, count)
  rand_cells <- c(rand_cells, rand_idx)
}
rna <- rna[,rand_cells]

#Protein
cur_cells <- sample(seq_len(ncol(protein)), 50000)
protein <- protein[,cur_cells]
```

# Find anchors

For mapping the two datasets have to be normalized in the same way and the used normalization method has to be passed in FindTransferAnchors!

Next, we will find anchors between the protein and RNA dataset.We can change the maximum allowed memory for future according to [this](https://satijalab.org/seurat/archive/v3.1/future_vignette.html).
```{r find-anchors, message=FALSE}
library(future)
library(parallel)
options(future.globals.maxSize= 1048576000)

plan("multisession", workers = detectCores() - 2)
anchors_protein <- FindTransferAnchors(reference = protein, query = rna, 
                                                  features = common_features,
                                                  reference.assay = "originalexp", 
                                                  query.assay = "RNA", 
                                                  reduction = "cca",
                                                  dims=1:20)

saveRDS(anchors_protein, file.path(path, "public_datasets/Jansky/anchors.rds"))

```

# Transfer labels

Next, we will transfer celltype labels from protein to RNA data, impute the protein expressions on the RNA datasets, and then project the RNA data onto the reference query.

```{r transfer-labels, message=FALSE}

# plan("multisession", workers = detectCores() - 2)
# rna_mapped <- MapQuery(
#     anchorset = anchors_protein, 
#     query = rna,
#     reference = protein, 
#     refdata = list(
#       celltype = protein@active.ident, 
#       imputed_protein = GetAssayData(protein, assay = "originalexp", slot = "data")[common_features, ]
#       ),
#     reference.reduction = "pca"#,
#     #reduction.model="umap.new"
#   )

#Alternatively: 
rna_mapped <- rna
plan("multisession", workers = detectCores() - 2)
imputed_protein <- TransferData(anchorset = anchors_protein,
                                 refdata = GetAssayData(protein, assay = "originalexp", slot = "data")[common_features, ],
                                 weight.reduction = rna[["pca"]],
                                 dims=1:20)

rna_mapped[["protein"]] <- imputed_protein

plan("multisession", workers = detectCores() - 2)
predicted_celltype <- TransferData(anchorset = anchors_protein,
                                               refdata = protein@active.ident,
                                               weight.reduction =rna[["pca"]],
                                               dims=1:20)

rna_mapped <- AddMetaData(rna_mapped, metadata = predicted_celltype[,c("predicted.id","prediction.score.max")])

saveRDS(protein, file.path(path, "public_datasets/Jansky/protein_mapped.rds"))
saveRDS(rna_mapped, file.path(path, "public_datasets/Jansky/rna_mapped.rds"))

rna_mapped@meta.data$predicted.id
unname(rna_mapped@active.ident)
```
# Explore results

Finally we will explore the mapping results.

```{r evaluate, fig.width=14, message=FALSE}
#DimPlot(rna_mapped, reduction = "ref.umap", group.by = "predicted.celltype.l1", label = TRUE, label.size = 3, repel = TRUE) + NoLegend()

library(RColorBrewer)
library(paletteer)


set.seed(20230521)

data <- data.frame(protein_clusters=rna_mapped@meta.data$predicted.id, rna_clusters=unname(rna_mapped@active.ident))

col_vector_433 = colors()[grep('gr(a|e)y', grDevices::colors(), invert = T)]
col_clusters = setNames(sample(col_vector_433, length(unique(data$rna_clusters))), unique(data$rna_clusters))

data$freq <- 1
data_wide <- data %>% pivot_wider(names_from = c("rna_clusters"), values_from = "freq", values_fn = sum, values_fill = 0)

#write.csv(data_wide, file.path(path, "/results/metadata/tissues-per-cluster.csv"))

data_long <- data_wide %>% pivot_longer(cols = !protein_clusters, names_to = "rna_clusters", values_to = "count") 

# Stacked + percent
ggplot(data_long, aes(fill=rna_clusters, y=count, x=protein_clusters)) + 
  ggtitle("RNA clusters per predicted protein cluster") +
  geom_bar(position="fill", stat="identity", show.legend = T) + 
  scale_fill_manual(values = col_clusters) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 
```