---
title: "Read in scRNAseq data"
output: html_document
date: '2023-07-06'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Read in the single-cell data {#read-data}

Here we will read in single-cell RNAseq data from a single-cell adrenal gland atlas by 
[Jansky et al.](https://adrenal.kitz-heidelberg.de/developmental_programs_NB_viz/), and
our imaging-based single-cell data

```{r read-data, message=FALSE}
library(Seurat)
library(SingleCellExperiment)
library(dplyr)
library(tidyr)
library(monocle3)

path <- "/mnt/Multimodal_Imaging_Daria/PIPELINE/NEWEST_RESULTS"

rna <- readRDS(file.path(path,"results/Manuscript/2_Figure/public_datasets/Lee/disco_bone_marrow_v2.0.rds"))
DefaultAssay(rna) <- "RNA"
rna@assays$RNA@data <- rna@assays$RNA@counts

protein <- readRDS(file.path(path,"results/Manuscript/spe_clustered.rds"))
protein <- protein[rowData(protein)$use_channel,]
col_celltype <- metadata(protein)$color_vectors$col_celltype
```

# Set common features 
Next, we will set common features between the reference and query dataset. Therefore,
we have to rename the features in the imaging dataset to the names of the corresponding
genes. 

```{r set common features, message=FALSE}

protein2gene <- read.csv(file.path(path,"results/public_datasets/protein2gene.csv"), sep=";")

code <- setNames(protein2gene$genes, protein2gene$protein)
rownames(protein) <- recode(rownames(protein), !!!code)
rowData(protein)$name <- rownames(protein)
rowData(protein)$channel <- rownames(protein)

protein <- as.Seurat(protein, counts = "counts", data = "exprs")
Idents(protein) <- "celltype"

#protein <- readRDS(file.path(path, "results/Manuscript/2_Figure/public_datasets/imc_seurat.rds"))

common_features <- intersect(rownames(rna@assays$RNA@data), row.names(protein))

```

# Quality control on public dataset

We will first look at cluster similarity to evaulate which clusters are similar. 
```{r mean-per-cluster-protein, fig.height=15, message=FALSE}
library(scuttle)
library(bruceR)
library(ComplexHeatmap)
library(viridis)

protein_clusters <- protein@active.ident
protein_data <- GetAssayData(protein, assay = "originalexp", slot = "data")[common_features, ]
protein_mean <- aggregate(t(protein_data), list(protein_clusters), mean)

protein_mean_t <- t(protein_mean[,2:ncol(protein_mean)])
colnames(protein_mean_t) <- protein_mean[,"Group.1"]
protein_mean_scaled <- t(apply(protein_mean_t, 1, RESCALE, to=0:1))

column_ha = HeatmapAnnotation(ncells=anno_barplot(as.data.frame(table(protein_clusters))[,"Freq"], height = unit(3, "cm")),
                              show_legend=F)

Heatmap(protein_mean_scaled, 
        column_title = "Mean per cluster (protein data)",
        top_annotation = column_ha, 
        cluster_columns=T,
        show_column_dend = T,
        show_column_names=T,
        show_row_names=T,
        col=viridis(100)
        )

```

```{r mean-per-cluster-rna, fig.height=15, message=FALSE}

#rna_clusters <- rna@meta.data$fate
rna_clusters <- rna@active.ident
rna_data <- GetAssayData(rna, assay = "RNA", slot = "data")[common_features, ]
rna_mean <- aggregate(t(rna_data), list(rna_clusters), mean)

rna_mean_t <- t(rna_mean[,2:ncol(rna_mean)])
colnames(rna_mean_t) <- rna_mean[,"Group.1"]
rna_mean_scaled <- t(apply(rna_mean_t, 1, RESCALE, to=0:1))

column_ha = HeatmapAnnotation(ncells=anno_barplot(as.data.frame(table(rna_clusters))[,"Freq"], height = unit(3, "cm")),
                              show_legend=F)

Heatmap(rna_mean_scaled, 
        column_title = "Mean per cluster (RNA data)",
        top_annotation = column_ha, 
        cluster_columns=T,
        show_column_dend = T,
        show_column_names=T,
        show_row_names=T,
        col=viridis(100)
        )

```
We will then visualize the two datasets by UMAP.

```{r umap-rna, fig.width=20, fig.height=15, message=FALSE}
library(RColorBrewer)
library(paletteer)
set.seed(230712)

col_vector_433 = colors()[grep('gr(a|e)y', grDevices::colors(), invert = T)]
col_rna_clusters = setNames(sample(col_vector_433, length(unique(rna@active.ident))), unique(rna@active.ident))

DimPlot(rna, reduction="umap", cols=col_rna_clusters)

```

```{r umap-protein, message=FALSE, fig.width=10, fig.height=7}
set.seed(230713)
rand_cells <- sample(seq_len(ncol(protein)), 10000)
DimPlot(protein[,rand_cells], reduction="UMAP", cols=col_celltype)
```

For further analysis we will exclude tumor cells and other.
```{r data-curation, message=FALSE}
protein <- protein[,protein@meta.data$metacluster!="tumor" & protein@meta.data$metacluster!="other"]
```

We will next plot the two together to look at similarity.
```{r mean-per-cluster-combined, fig.height=12, fig.width=16, message=FALSE}
set.seed(230717)
protein_clusters <- protein@active.ident
protein_data <- GetAssayData(protein, assay = "originalexp", slot = "data")[common_features, ]
protein_mean <- aggregate(t(protein_data), list(protein_clusters), mean)
protein_mean_t <- t(protein_mean[,2:ncol(protein_mean)])
colnames(protein_mean_t) <- protein_mean[,"Group.1"]
protein_mean_scaled <- t(apply(protein_mean_t, 1, RESCALE, to=0:1))

rna_clusters <- droplevels(rna@active.ident)
rna_data <- GetAssayData(rna, assay = "RNA", slot = "data")[common_features, ]
rna_mean <- aggregate(t(rna_data), list(rna_clusters), mean)
rna_mean_t <- t(rna_mean[,2:ncol(rna_mean)])
colnames(rna_mean_t) <- rna_mean[,"Group.1"]
rna_mean_scaled <- t(apply(rna_mean_t, 1, RESCALE, to=0:1))

clusters <- c(rna_clusters, protein_clusters)
cell_count <- c(as.data.frame(table(rna_clusters))[,"Freq"], as.data.frame(table(protein_clusters))[,"Freq"])
modality <- c(rep(c("RNA"), ncol(rna_mean_scaled)), rep(c("protein"), ncol(protein_mean_scaled)))
mean_scaled <- cbind(rna_mean_scaled, protein_mean_scaled)

column_ha = HeatmapAnnotation(modality=modality,
                              ncells=anno_barplot(cell_count, height = unit(3, "cm")),
                              show_legend=F)

Heatmap(mean_scaled, 
        column_title = "Mean per cluster (RNA & protein)",
        top_annotation = column_ha, 
        cluster_columns=T,
        show_column_dend = T,
        show_column_names=T,
        show_row_names=T,
        col=viridis(100)
        )


```

We will then look at cluster similarity in the PCA plot.
```{r pca-plot, fig.width=20, message=FALSE}
library("FactoMineR")
library("factoextra")
library("tidyr")

data <- as.data.frame(mean_scaled)

data <- t(data)
data.pca <- FactoMineR::PCA(data, graph=F)
data$modality <- as.factor(modality)

fviz_pca_ind(data.pca, repel=T, habillage=data$modality)
```

#Transfer labels with singleR

Finally, we will test singleR to transfer labels from the RNA to the protein dataset. 

```{r single-r, fig.height=13, fig.width= 20, message=FALSE}
library(BiocParallel)
library(SingleR)

rna_mat <- GetAssayData(rna, assay = "RNA", slot = "data")[common_features, ]
rna_clusters <- rna@active.ident
protein_mat <- GetAssayData(protein, assay = "originalexp", slot = "counts")[common_features, ]

rm(rna)
rm(protein)

predicted_labels <- SingleR(test = protein_mat, ref = rna_mat, labels = rna_clusters, BPPARAM = MulticoreParam(workers=16, RNGseed=20231030))
saveRDS(predicted_labels, file.path(path, "results/Manuscript/2_Figure/public_datasets/Lee/predicted_labels.rds"))

#predicted_labels <- readRDS(file.path(path, "results/Manuscript/2_Figure/public_datasets/Lee/predicted_labels.rds"))

setEPS()
postscript(file.path(path, "results/Manuscript/2_Figure/public_datasets/Lee/score_heatmap.eps"), height=8, width=12)
plotScoreHeatmap(predicted_labels[sample(seq_len(nrow(predicted_labels)), 16000),],annotation_colors=list(Labels = col_clusters), annotation_legend=F, treeheight_row=0) #define colors first
dev.off()

tiff(filename = file.path(path, "results/Manuscript/2_Figure/public_datasets/Lee/score_heatmap.tiff"),
   width = 90, height = 70, units = "mm", res=600)
plotScoreHeatmap(predicted_labels[sample(seq_len(nrow(predicted_labels)), 16000),],annotation_colors=list(Labels = col_clusters), annotation_legend=F, treeheight_row=0) #define colors first
dev.off()
```

Next, we will look at the marker expression in a single-cell heatmap and plot the true and predicted labels in the top annotation bar.

```{r heatmap prediciton, fig.height=5, fig.width= 8, message=FALSE}
library(RColorBrewer)
set.seed(20230517)

true_label <- protein@active.ident
predicted_label <- predicted_labels$labels

n <- length(unique(predicted_label))

qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector_74 = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
col_vector_433 = colors()[grep('gr(a|e)y', grDevices::colors(), invert = T)]

col_predicted_labels = setNames(sample(col_vector_433, length(unique(predicted_label))), unique(predicted_label))
col_true_labels = col_celltype

count <- min(table(predicted_label))
count <- 500
rand_cells <- c()

#Get same number of PT and BM cells per cluster
for (i in 1: length(unique(predicted_label))){
  idx <- which(predicted_label == unique(predicted_label)[i])
  rand_idx <- sample(idx, count)
  rand_cells <- c(rand_cells, rand_idx)
}

rand_cells <- sample(seq_len(ncol(rna_mat)), 4800)

column_ha = HeatmapAnnotation(predicted_label = predicted_label[rand_cells],
                              true_cluster = true_label[rand_cells],
                              col=list(predicted_label=col_predicted_labels, 
                                       true_label=col_true_labels
                                       ), 
                              show_legend=T)


protein_mat_scaled <- t(apply(protein_mat[,rand_cells], 1, RESCALE, to=0:1))

Heatmap(as.matrix(protein_mat_scaled), 
        column_title = "Protein data- scHeatmap - true and predicted labels",
        top_annotation = column_ha, 
        cluster_columns=F,
        column_order= order(predicted_label[rand_cells], decreasing=F),
        show_column_dend = T,
        show_column_names=F,
        col=viridis(100)
        )
```

```{r mean-per-cluster-protein, fig.height= 15, fig.width= 15, message=FALSE}
library(bruceR)

mat <- protein_mat

mean <- aggregate(t(mat), list(predicted_labels$labels), mean)

mean_t <- t(mean[,2:ncol(mean)])
colnames(mean_t) <- mean[,"Group.1"]
mean_scaled <- t(apply(mean_t, 1, RESCALE, to=0:1))

column_ha = HeatmapAnnotation(ncells=anno_barplot(as.data.frame(table(predicted_labels$labels))[,"Freq"], height = unit(1, "cm")), 
                              show_legend=F)

#tumor_markers <- c("ELAVL4", "CXCR4", "MME", "CD74", "HLA-A", "CD274", "GATA3", "VIM", "MKI67", "CD44", "NCAM1", "ST8SIA1", "PRPH", "LUM", "CHGA", "SOX10", "S100B")


Heatmap(mean_scaled[common_features,], 
        column_title = "Mean per cluster (protein data)",
        top_annotation = column_ha, 
        cluster_columns=T,
        show_column_dend = T,
        show_column_names=T,
        show_row_names=T,
        col=viridis(100)
        )

```

Finally we will explore the mapping results in a barplot.

```{r evaluate-barplot, fig.width=16, message=FALSE}

data <- data.frame(RNA_clusters=predicted_label, protein_clusters=protein@active.ident)

data$freq <- 1
data_wide <- data %>% pivot_wider(names_from = c("protein_clusters"), values_from = "freq", values_fn = sum, values_fill = 0)

data_long <- data_wide %>% pivot_longer(cols = !RNA_clusters, names_to = "protein_clusters", values_to = "count") 

# Stacked + percent
ggplot(data_long, aes(fill=RNA_clusters, y=count, x=protein_clusters)) + 
  ggtitle("Predicted RNA cluster per protein cluster") +
  geom_bar(position="fill", stat="identity", show.legend = T) + 
  scale_fill_manual(values = col_predicted_labels) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 
```
Next, for each protein based cell type, we will determine the major RNA celltype assignment.
```{r evaluate-max-assignment, fig.width=20, message=FALSE}
RNA_clusters <- data_wide$RNA_clusters
data <- data_wide
data <- data[,-1]
rownames(data) <- RNA_clusters
data <- (prop.table(as.matrix(data), margin=2))
protein_clust <- names(unlist(apply(data, 2, which.max)))
predicted_rna_clust_1 <- data_wide[unname(unlist(apply(data, 2, which.max))), "RNA_clusters"]
proportion_1 <- unname(unlist(apply(data, 2, max)))
predicted_rna_clust_2 <- unname(apply(data, 2, function(x){rownames(data)[order(x, decreasing=T)][2]}))
proportion_2 <- unname(apply(data, 2, function(x){sort(x, decreasing=T)[2]}))

assignments <- as.data.frame(cbind("protein_cluster" = protein_clust, "predicted_RNA_cluster_1" = predicted_rna_clust_1, "proportion_1"=proportion_1, "predicted_RNA_cluster_2" = predicted_rna_clust_2, "proportion_2"=proportion_2))

assignments
```
Next, we will save predicted clusters in a separate slot in the spe object and define colors.

```{r save-predicted, message=FALSE, fig.height=17}
spe <- readRDS(file.path(path,"results/Manuscript/spe_clustered.rds"))
spe$celltype_predicted <- "tumor"
spe[,spe$metacluster=="other"]$celltype_predicted <- "other"
spe[,rownames(predicted_labels)]$celltype_predicted <- predicted_labels$labels


predicted_colors <- read.csv(file.path(path, "results/Manuscript/2_Figure/public_datasets/Lee/celltype_metacluster.csv"), sep=";")
col_predicted <- setNames(predicted_colors$color, predicted_colors$celltype)
col_predicted <- c(col_predicted, c("tumor" = "snow2", "other"="goldenrod1"))

pie(rep(1, length(col_predicted)),  col=col_predicted, labels=names(col_predicted))

metadata(spe)$color_vectors$col_predicted_label <- col_predicted

saveRDS(spe, file.path(path, "results/Manuscript/2_Figure/public_datasets/Lee/spe_predicted.rds"))
```

We will plot the predicted labels in a UMAP.

```{r cell type umap, fig.width=15, fig.height=15, message=FALSE}
library(dittoSeq)
library(scater)
library(patchwork)
library(cowplot)
library(viridis)


## UMAP colored by cell type and expression - dittoDimPlot
p1 <- dittoDimPlot(spe, var = "celltype_predicted", 
             reduction.use = "UMAP", size = 0.3, labels.size=2,
             do.label = F,
             legend.show=F) +
  scale_color_manual(values = metadata(spe)$color_vectors$col_predicted_label) +
  theme(legend.title = element_blank()) +
  ggtitle("Celltype on UMAP")

tiff(filename = file.path(path, "results/Manuscript/2_Figure/public_datasets/Lee/UMAP_predicted_celltype2_tumor-white.tiff"),
     width = 115, height = 95, units = "mm", res=600)
p1
dev.off()


```

We will then save the spe object. 

```{r save-spe, message=FALSE}

saveRDS(spe, file.path(path, "results/Manuscript/2_Figure/public_datasets/Lee/spe_predicted.rds"))
```
