<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="LazDaria" />

<meta name="date" content="2024-03-25" />

<title>Read single-cell and meta-data</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">MapMetSC</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Read single-cell and meta-data</h1>
<h4 class="author">LazDaria</h4>
<h4 class="date">2024-03-25</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-03-25
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 5
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 2
</p>
<p>
<strong>Knit directory:</strong> <code>MapMetSC/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R
Markdown file created these results, you’ll want to first commit it to
the Git repo. If you’re still working on the analysis, you can ignore
this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240314code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240314)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240314code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240314)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongdetected">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>Cache:</strong> detected </a>
</p>
</div>
<div id="strongCachestrongdetected" class="panel-collapse collapse">
<div class="panel-body">
The following chunks had caches available:
<ul>
<li>
read-steinbock
</li>
</ul>
<p>To ensure reproducibility of the results, delete the cache directory
<code>01_read_data_cache</code> and re-run the analysis. To have
workflowr automatically delete the cache directory prior to building the
file, set <code>delete_cache = TRUE</code> when running
<code>wflow_build()</code> or <code>wflow_publish()</code>.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomTaschnerMandlGroupMapMettreec2d11767240409e6b8384dbc6b6bef78127f3e2dtargetblankc2d1176a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/TaschnerMandlGroup/MapMet/tree/c2d11767240409e6b8384dbc6b6bef78127f3e2d" target="_blank">c2d1176</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomTaschnerMandlGroupMapMettreec2d11767240409e6b8384dbc6b6bef78127f3e2dtargetblankc2d1176a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/TaschnerMandlGroup/MapMet/tree/c2d11767240409e6b8384dbc6b6bef78127f3e2d" target="_blank">c2d1176</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    analysis/01_read_data_cache/

Untracked files:
    Untracked:  analysis/01_read_data.Rmd
    Untracked:  analysis/02_QC_1.Rmd
    Untracked:  analysis/style.css

Unstaged changes:
    Deleted:    R/20231214_Manuscript/1_Ext_Figure/01_batcheffect.Rmd
    Deleted:    R/20231214_Manuscript/1_Figure/01_read_single-cell-data_0px.Rmd
    Deleted:    R/20231214_Manuscript/1_Figure/02_read_single-cell-data_1px.Rmd
    Deleted:    R/20231214_Manuscript/1_Figure/03_removal_lost_cells.Rmd
    Deleted:    R/20231214_Manuscript/1_Figure/04_integration_morph-features.Rmd
    Deleted:    R/20231214_Manuscript/1_Figure/05_cell_phenotyping.Rmd
    Deleted:    R/20231214_Manuscript/1_Figure/06_vis_annotated_celltype.Rmd
    Deleted:    R/20231214_Manuscript/1_Figure/07_visualization_2.Rmd
    Deleted:    R/20231214_Manuscript/1_Figure/08_singleR_Fetahu.Rmd
    Deleted:    R/20231214_Manuscript/1_Figure/09_cohort_overview.Rmd
    Deleted:    R/20231214_Manuscript/1_Figure/Suppl_Table1.Rmd
    Deleted:    R/20231214_Manuscript/1_Figure/Suppl_Table2.Rmd
    Deleted:    R/20231214_Manuscript/2_Ext_Figure/01_bubble_plots.Rmd
    Deleted:    R/20231214_Manuscript/2_Figure/08_singleR_Fetahu.Rmd
    Deleted:    R/20231214_Manuscript/2_Figure/08_singleR_Jansky.Rmd
    Deleted:    R/20231214_Manuscript/2_Figure/08_singleR_Lee.Rmd
    Deleted:    R/20231214_Manuscript/3_Figure/09_DA_limma.Rmd
    Deleted:    R/20231214_Manuscript/3_Figure/9_DA_cellular_communities.Rmd
    Deleted:    R/20231214_Manuscript/4_Figure/09_DA_limma_PTvsBM.Rmd
    Deleted:    R/20231214_Manuscript/4_Figure/09_DA_limma_crtlvsinf.Rmd
    Deleted:    R/20231214_Manuscript/5_Figure/09_DA_cellular_communities_BM-DE.Rmd
    Deleted:    R/20231214_Manuscript/5_Figure/09_DA_limma_BM_DE.Rmd
    Deleted:    R/20231214_Manuscript/6_Figure/09_DA_cellular_communities_BM_timepoint.Rmd
    Deleted:    R/20231214_Manuscript/6_Figure/09_DA_cellular_communities_BM_timepoint_sepclustering.Rmd
    Deleted:    R/20231214_Manuscript/6_Figure/09_DA_limma_BM_timepoint.Rmd
    Deleted:    R/IMCData_analysis/01_read_transform_data.Rmd
    Deleted:    R/IMCData_analysis/01_read_transform_data.html
    Deleted:    R/IMCData_analysis/02_generate_spillovermatrix.Rmd
    Deleted:    R/IMCData_analysis/02_spillover_compensation.Rmd
    Deleted:    R/IMCData_analysis/02_spillover_correction.Rmd
    Deleted:    R/IMCData_analysis/02_spillover_correction.html
    Deleted:    R/IMCData_analysis/02_spillover_correction_imagebased.Rmd
    Deleted:    R/IMCData_analysis/03_quality_control.Rmd
    Deleted:    R/IMCData_analysis/03_quality_control.html
    Deleted:    R/IMCData_analysis/04_batch_correction_fullDataset.Rmd
    Deleted:    R/IMCData_analysis/04_batch_correction_fullDataset.html
    Deleted:    R/IMCData_analysis/05_clustering_fullDataset.Rmd
    Deleted:    R/IMCData_analysis/05_clustering_fullDataset.html
    Deleted:    R/IMCData_analysis/06_data_curation.Rmd
    Deleted:    R/IMCData_analysis/07_batch_correction_curatedDataset.Rmd
    Deleted:    R/IMCData_analysis/08_clustering_curatedDataset.Rmd
    Deleted:    R/IMCData_analysis/08_compare_clustering_methods.Rmd
    Deleted:    R/IMCData_analysis/09_image_visualization.Rmd
    Deleted:    R/IMCData_analysis/10_sc_visualization.Rmd
    Deleted:    R/IMCData_analysis/11_pseudobulk_analysis.Rmd
    Deleted:    R/IMCData_analysis/cluster_post-curation.Rmd
    Deleted:    R/IMCData_analysis/phenotyping_old.html
    Deleted:    R/IMCData_analysis/removal_lost_cells.Rmd
    Deleted:    R/IMCData_analysis/visualize_metadata.Rmd
    Deleted:    R/IMCData_analysis/visualize_metadata.html
    Deleted:    R/IMCData_analysis_2/01_read_transform_data.Rmd
    Deleted:    R/IMCData_analysis_2/02_removal_lost_cells.Rmd
    Deleted:    R/IMCData_analysis_2/03_integration_morph-features.Rmd
    Deleted:    R/IMCData_analysis_2/04_cell_phenotyping.Rmd
    Deleted:    R/IMCData_analysis_2/05_vis_annotated_celltype.Rmd
    Deleted:    R/IMCData_analysis_2/06_visualization_2.Rmd
    Deleted:    R/IMCData_analysis_2/07_visualize_metadata.Rmd
    Deleted:    R/IMCData_analysis_2/old_versions/04_cell_phenotyping.Rmd
    Deleted:    R/IMData_analysis_3/01_read_single-cell-data_0px.Rmd
    Deleted:    R/IMData_analysis_3/02_read_single-cell-data_1px.Rmd
    Deleted:    R/IMData_analysis_3/03_removal_lost_cells.Rmd
    Deleted:    R/IMData_analysis_3/04_integration_morph-features.Rmd
    Deleted:    R/IMData_analysis_3/05_02_cell_phenotyping.Rmd
    Deleted:    R/IMData_analysis_3/05_cell_phenotyping.Rmd
    Deleted:    R/IMData_analysis_3/06_metacluster_subphenotyping.Rmd
    Deleted:    R/IMData_analysis_3/06_vis_annotated_celltype.Rmd
    Deleted:    R/IMData_analysis_3/07_visualization_2.Rmd
    Deleted:    R/IMData_analysis_3/08_visualize_metadata.Rmd
    Deleted:    R/IMData_analysis_3/09_mm_mapping.Rmd
    Deleted:    R/IMData_analysis_3/09_mm_mapping2.Rmd
    Deleted:    R/IMData_analysis_3/09_mm_mapping_vignette.Rmd
    Deleted:    R/IMData_analysis_3/09_singleR_RNA.Rmd
    Deleted:    R/IMData_analysis_3/09_singleR_protein.Rmd
    Deleted:    R/IMData_analysis_3/10_visualize_predicted_clusters_per_metadata.Rmd
    Deleted:    R/IMData_analysis_3/11_DESeq2.Rmd
    Deleted:    R/IMData_analysis_3/11_cyto_diffcyt.Rmd
    Deleted:    R/IMData_analysis_3/11_cytofworkflow_DME.Rmd
    Deleted:    R/IMData_analysis_3/11_cytofworkflow_diffcyt.Rmd
    Deleted:    R/IMData_analysis_3/11_diffcyt.Rmd
    Deleted:    R/IMData_analysis_3/11_diffcyt_limma-voom.Rmd
    Deleted:    R/IMData_analysis_3/11_edgeR.Rmd
    Deleted:    R/IMData_analysis_3/11_miloR.Rmd
    Deleted:    R/IMData_analysis_3/11_miloR_longitudinal.Rmd
    Deleted:    R/IMData_analysis_3/11_milo_DA.Rmd
    Deleted:    R/IMData_analysis_3/12_pseudotime_scorpion_in-house.Rmd
    Deleted:    R/IMData_analysis_3/12_pseudotime_scorpion_vignette.Rmd
    Deleted:    R/IMData_analysis_3/CytoGLMM.Rmd
    Deleted:    R/IMData_analysis_3/PBMC8_cluster_merging1.xlsx
    Deleted:    R/IMData_analysis_3/PBMC8_cluster_merging2_v3.xlsx
    Deleted:    R/IMData_analysis_3/PBMC8_metadata.xlsx
    Deleted:    R/IMData_analysis_3/PBMC8_panel_v3.xlsx
    Deleted:    R/IMData_analysis_3/diffcyt.Rmd
    Deleted:    R/workflowR.Rmd
    Modified:   analysis/_site.yml
    Deleted:    analysis/first-analysis.Rmd

Staged changes:
    New:        code/R/20231214_Manuscript/1_Figure/01_read_single-cell-data_0px.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with
<code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<div id="read-data" class="section level1">
<h1>Read in the data</h1>
<p>We will first read in single-cell data, previously extracted from
processed multi-modal (IMC,IF) images in <a
href="https://github.com/TaschnerMandlGroup/MapMetIP">MapMetIP</a>. For
reading the data, we use the <a
href="https://github.com/BodenmillerGroup/imcRtools">imcRtools</a>
package. This package uses the <code>read_steinbock</code> function to
read data into a <code>SpatialExperiment</code>. The <a
href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html">SpatialExperiment</a>
class <span class="citation">[@Righelli2021]</span> is an extension of
the <code>SingleCellExperiment</code> class. For more information,
please refer to <code>?read_steinbock</code>.</p>
<div id="load-libraries" class="section level2">
<h2>Load libraries</h2>
<details>
<summary>
Load libraries
</summary>
<pre class="r"><code>library(imcRtools)</code></pre>
<pre><code>Loading required package: SpatialExperiment</code></pre>
<pre><code>Loading required package: SingleCellExperiment</code></pre>
<pre><code>Loading required package: SummarizedExperiment</code></pre>
<pre><code>Loading required package: MatrixGenerics</code></pre>
<pre><code>Loading required package: matrixStats</code></pre>
<pre><code>
Attaching package: &#39;MatrixGenerics&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:matrixStats&#39;:

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars</code></pre>
<pre><code>Loading required package: GenomicRanges</code></pre>
<pre><code>Loading required package: stats4</code></pre>
<pre><code>Loading required package: BiocGenerics</code></pre>
<pre><code>
Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:stats&#39;:

    IQR, mad, sd, var, xtabs</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    anyDuplicated, append, as.data.frame, basename, cbind, colnames,
    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
    union, unique, unsplit, which.max, which.min</code></pre>
<pre><code>Loading required package: S4Vectors</code></pre>
<pre><code>
Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    expand.grid, I, unname</code></pre>
<pre><code>Loading required package: IRanges</code></pre>
<pre><code>Loading required package: GenomeInfoDb</code></pre>
<pre><code>Loading required package: Biobase</code></pre>
<pre><code>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    &#39;browseVignettes()&#39;. To cite Bioconductor, see
    &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<pre><code>
Attaching package: &#39;Biobase&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:MatrixGenerics&#39;:

    rowMedians</code></pre>
<pre><code>The following objects are masked from &#39;package:matrixStats&#39;:

    anyMissing, rowMedians</code></pre>
<pre class="r"><code>library(stringr)
library(RColorBrewer)
library(circlize)</code></pre>
<pre><code>========================================
circlize version 0.4.15
CRAN page: https://cran.r-project.org/package=circlize
Github page: https://github.com/jokergoo/circlize
Documentation: https://jokergoo.github.io/circlize_book/book/

If you use it in published research, please cite:
Gu, Z. circlize implements and enhances circular visualization
  in R. Bioinformatics 2014.

This message can be suppressed by:
  suppressPackageStartupMessages(library(circlize))
========================================</code></pre>
<pre class="r"><code>library(paletteer)
library(dittoSeq)</code></pre>
<pre><code>Loading required package: ggplot2</code></pre>
</details>
</div>
<div id="read-single-cell-measurement" class="section level2">
<h2>Read single-cell measurement</h2>
<p>Replace <code>path</code> with the path to the folder where you
stored the single-cell data. In <a
href="https://github.com/TaschnerMandlGroup/MapMetIP">MapMetIP</a>, we
exported single-cell marker measurements based on non-dilated and
dilated (by 1px) segmentation masks to capture membrane signals. As both
are required in the downstream analysis, we will read in both and
combine them into one SPE object saving each expression matrix into one
<code>assays</code>slot. Measurements based on non-dilated masks will be
in the <code>assays</code>slot <code>counts</code> and the measurements
based on dilated masks will be stored in <code>counts_1px</code>.</p>
<pre class="r"><code>path &lt;- params$input

spe_0px &lt;- read_steinbock(path=path, image_file=NULL, panel_file = file.path(path, &#39;NB_Panel.csv&#39;), intensities_folder = &quot;intensities-0px&quot;,regionprops_folder = &quot;regionprops&quot;, graphs_folder=NULL) 
spe_1px &lt;- read_steinbock(path=path, image_file=NULL, panel_file = file.path(path, &#39;NB_Panel.csv&#39;), intensities_folder = &quot;intensities-1px&quot;,regionprops_folder = &quot;regionprops&quot;, graphs_folder=NULL) 

#Combine SPE objects into one object
spe &lt;- spe_0px
assay(spe, &quot;counts_1px&quot;) &lt;- assay(spe_1px, &quot;counts&quot;)

#Rename third IF channel (CD45 for PT, CD56 for BM)
rownames(spe)[44] &lt;- &quot;IF3_CD45/56_mean&quot;
rownames(spe)[88] &lt;- &quot;IF3_CD45/56_mean-80&quot;</code></pre>
<div class="alert alert-warning">
<p>
<strong>Warning:</strong> The above code chunk cached its results, but
it won’t be re-run if previous chunks it depends on are updated. If you
need to use caching, it is highly recommended to also set
<code>knitr::opts_chunk$set(autodep = TRUE)</code> at the top of the
file (in a chunk that is not cached). Alternatively, you can customize
the option <code>dependson</code> for each individual chunk that is
cached. Using either <code>autodep</code> or <code>dependson</code> will
remove this warning. See the
<a href="https://yihui.name/knitr/options/#cache"
>knitr cache options</a> for more details.
</p>
</div>
</div>
<div id="read-single-cell-metadata" class="section level2">
<h2>Read single-cell metadata</h2>
<p>We set the <code>colnames</code> of the object to generate unique
identifiers per cell.</p>
<pre class="r"><code>colnames(spe) &lt;- paste0(spe$sample_id, &quot;_&quot;, spe$ObjectNumber)</code></pre>
<p>When then read sample and patient-specific metadata from an external
file.</p>
<pre class="r"><code>spe$sample &lt;- gsub(spe$sample_id, pattern=&quot;([UM]).*&quot;, replacement=&quot;\\1&quot;)
spe$fm_id &lt;- str_match(spe$sample, &quot;_.*_(.*)_&quot;)[,2]  
spe$tissue &lt;- str_match(spe$sample, &quot;^.*_.*_(.*)&quot;)[,2] 
spe$tissue[spe$tissue==&quot;TU&quot;] = &quot;PT&quot;

meta &lt;- read.csv(file.path(path, &#39;metadata_complete_anon.csv&#39;))

spe$puncture_side &lt;- meta$Puncture_side[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$infiltration_right &lt;- meta$Infiltration_right[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$infiltration_left &lt;- meta$Infiltration_left[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$exclude &lt;- meta$exclude[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$imc_score &lt;- meta$IMC_score[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$staining_date &lt;- meta$Staining_date[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$stainer &lt;- meta$Stainer[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$imc_date &lt;- meta$IMC_date[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$DE_date &lt;- meta$DE_date[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$DE_year &lt;- sub(&quot;-.*&quot;, &quot;&quot;, spe$DE_date)
spe$control &lt;- meta$control[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$study_id &lt;- meta$Study_ID[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$timepoint &lt;- meta$Timepoint[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$stage &lt;- meta$stage4[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$progression &lt;- meta$efs[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$tod &lt;- meta$tod[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$efs_dur &lt;- meta$efs_dur[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$tod_date &lt;- meta$tod_dat1[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$tod_dur &lt;- meta$tod_dur[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$rez_date &lt;- meta$rez_dat[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$prog_date &lt;- meta$prog_dat[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$event_date &lt;- meta$event_dat[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$X1p_loss &lt;- meta$X1p_loss[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$X1q_gain &lt;- meta$X1q_gain[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$X1q_loss &lt;- meta$X1q_loss[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$X2p_gain &lt;- meta$X2p_gain[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$X3p_loss &lt;- meta$X3p_loss[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$X3q_loss &lt;- meta$X3q_loss[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$X4p_loss &lt;- meta$X4p_loss[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$X6q_loss &lt;- meta$X6q_loss[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$X7q_gain &lt;- meta$X7q_gain[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$PTPRD_del &lt;- meta$PTPRD_del[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$CDKN2AB_del &lt;- meta$CDKN2AB_del[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$X11q_loss &lt;- meta$X11q_loss[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$X12q_gain &lt;- meta$X12q_gain[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$X14q_loss &lt;- meta$X14q_loss[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$X17p_loss &lt;- meta$X17p_loss[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$X17q_gain &lt;- meta$X17q_gain[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$X19p_loss &lt;- meta$X19p_loss[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$X19q_loss &lt;- meta$X19q_loss[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$X22q_loss &lt;- meta$X22q_loss[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$MYCN_amp &lt;- meta$MYCN_amp[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$TERT_rear &lt;- meta$TERT_rear[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$ATRX_del &lt;- meta$ATRX_del[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$ALK_amp &lt;- meta$ALK_amp[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$ALK_mut &lt;- meta$ALK_mut[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]

spe$age_diag &lt;- meta$age_at_diag[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]
spe$therapy &lt;- meta$therapy[match(paste(spe$fm_id, spe$tissue), paste(meta$Sample, meta$Tissue))]

spe$ROI &lt;- str_match(spe$sample_id, &quot;^.*_.*_.*_.*_(.*)&quot;)[,2]
spe$year &lt;- gsub(spe$sample_id, pattern=&quot;^.*_.*_(.*)-.*&quot;, replacement=&quot;\\1&quot;)</code></pre>
<p>We assign color codes to all metadata variables for later
visualizations.</p>
<pre class="r"><code>#Get a named vector a different colors
qual_col_pals = brewer.pal.info[brewer.pal.info$category == &#39;qual&#39;,]
col_vector_74 = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
col_vector_433 = colors()[grep(&#39;gr(a|e)y&#39;, grDevices::colors(), invert = T)]

#Assign colors to metavariables
meta_all &lt;- meta
meta_all &lt;- meta_all[order(meta_all$Study_ID),] #order by study-id

exclude &lt;- replace(x = meta_all$exclude, list =  !meta_all$exclude %in% c(&#39;no&#39;, &#39;maybe&#39;), values =  &#39;partially&#39;)

DE_year &lt;- sub(&quot;-.*&quot;, &quot;&quot;, meta_all$DE_date)

metadata(spe)$color_vectors$col_sample = setNames(sample(col_vector_433, length(unique(meta_all$Sample))), unique(meta_all$Sample))
metadata(spe)$color_vectors$col_puncture_side = setNames(brewer.pal(length(unique(meta_all$Puncture_side)), &#39;Set3&#39;), unique(meta_all$Puncture_side))
metadata(spe)$color_vectors$col_staining_batch = setNames(sample(col_vector_74, length(unique(meta_all$Staining_date))), unique(meta_all$Staining_date))
metadata(spe)$color_vectors$col_stainer = setNames(c(&#39;#a6611a&#39;, &#39;#dfc27d&#39;, &#39;#80cdc1&#39;, &#39;#018571&#39;), unique(meta_all$Stainer))
metadata(spe)$color_vectors$col_imc_batch = setNames(sample(col_vector_74, length(unique(meta_all$IMC_date))), unique(meta_all$IMC_date))
metadata(spe)$color_vectors$col_imc_score = colorRamp2(c(min(meta_all$IMC_score, na.rm = TRUE), max(meta_all$IMC_score, na.rm = TRUE)/2, max(meta_all$IMC_score, na.rm = TRUE)), c(&quot;red&quot;, &quot;white&quot;, &quot;blue&quot;)) 
metadata(spe)$color_vectors$col_study_id = setNames(sample(col_vector_433, length(unique(meta_all$Study_ID))), unique(meta_all$Study_ID))
metadata(spe)$color_vectors$col_DE_year = setNames(sample(col_vector_74, length(unique(DE_year))), unique(DE_year))
metadata(spe)$color_vectors$col_timepoint = setNames(c(&#39;#7fc97f&#39;, &#39;#beaed4&#39;, &#39;#fdc086&#39; , &#39;#ffff99&#39;, &#39;#386cb0&#39;), unique(meta_all$Timepoint))
metadata(spe)$color_vectors$col_aberrations = setNames(c(&#39;#018571&#39;,&#39;#a6611a&#39;,&#39;#80cdc1&#39;,&#39;#f5f5f5&#39;), c(0,1,3,9))
metadata(spe)$color_vectors$col_stages = setNames(c(&#39;#a6cee3&#39;, &#39;#1f78b4&#39;, &#39;#b2df8a&#39;, &#39;#33a02c&#39;, &#39;#fb9a99&#39;, &#39;#e31a1c&#39;), unique(meta_all$stage4))
metadata(spe)$color_vectors$col_efs_dur = colorRamp2(c(min(meta_all$efs_dur, na.rm = TRUE), max(meta_all$efs_dur, na.rm = TRUE)/2, max(meta_all$efs_dur, na.rm = TRUE)), paletteer_c(&quot;ggthemes::Red-Green Diverging&quot;, 3))
metadata(spe)$color_vectors$col_tissue = setNames(c(&#39;aquamarine4&#39;,&#39;brown&#39;), unique(meta_all$Tissue))
metadata(spe)$color_vectors$col_control = setNames(c(&#39;mediumseagreen&#39;, &#39;gray77&#39;), unique(meta_all$control))
metadata(spe)$color_vectors$col_exclude = setNames(c(&#39;#018571&#39;,&#39;#a6611a&#39;,&#39;#80cdc1&#39;), unique(exclude))
metadata(spe)$color_vectors$col_progress = setNames(c(&#39;gray77&#39;, &#39;palegreen3&#39;, &#39;firebrick1&#39;), c(&#39;n.a.&#39;, 0, 1))</code></pre>
<p>Before we save the spatial experiment object, we check the
distribution of markers.</p>
<pre class="r"><code>marker &lt;- &#39;CD3_Sm152_mean&#39;
tissue &lt;- unique(spe$tissue)

dittoRidgePlot(spe[,spe$tissue%in%tissue], var = marker, group.by = &quot;tissue&quot;, assay = &quot;counts&quot;) +
    ggtitle(paste(marker,  &quot;- per tissue&quot;))</code></pre>
<pre><code>Picking joint bandwidth of 0.0173</code></pre>
<p><img src="figure/01_read_data.Rmd/plot%20by%20staining%20date-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="save-object" class="section level2">
<h2>Save object</h2>
<p>Finally, the generated data objects can be saved for further
downstream processing and analysis.</p>
<pre class="r"><code>assay(spe, &quot;exprs&quot;) &lt;- assay(spe, &quot;counts&quot;)

saveRDS(spe, file.path(params$output, &quot;spe.rds&quot;))</code></pre>
</div>
<div id="session-info" class="section level2">
<h2>Session Info</h2>
<details>
<summary>
SessionInfo
</summary>
<pre><code>R version 4.2.0 (2022-04-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] dittoSeq_1.8.1              ggplot2_3.4.4              
 [3] paletteer_1.5.0             circlize_0.4.15            
 [5] RColorBrewer_1.1-3          stringr_1.5.1              
 [7] imcRtools_1.3.5             SpatialExperiment_1.6.1    
 [9] SingleCellExperiment_1.20.1 SummarizedExperiment_1.26.1
[11] Biobase_2.56.0              GenomicRanges_1.48.0       
[13] GenomeInfoDb_1.32.4         IRanges_2.30.1             
[15] S4Vectors_0.34.0            BiocGenerics_0.42.0        
[17] MatrixGenerics_1.8.1        matrixStats_1.1.0          
[19] workflowr_1.7.1            

loaded via a namespace (and not attached):
  [1] utf8_1.2.4                shinydashboard_0.7.2     
  [3] R.utils_2.12.3            tidyselect_1.2.0         
  [5] htmlwidgets_1.6.4         grid_4.2.0               
  [7] BiocParallel_1.30.4       DropletUtils_1.16.0      
  [9] munsell_0.5.0             codetools_0.2-18         
 [11] units_0.8-5               DT_0.31                  
 [13] withr_3.0.0               colorspace_2.1-0         
 [15] highr_0.10                knitr_1.45               
 [17] rstudioapi_0.15.0         labeling_0.4.3           
 [19] git2r_0.33.0              GenomeInfoDbData_1.2.8   
 [21] polyclip_1.10-6           bit64_4.0.5              
 [23] farver_2.1.1              pheatmap_1.0.12          
 [25] rhdf5_2.40.0              rprojroot_2.0.4          
 [27] vctrs_0.6.5               generics_0.1.3           
 [29] xfun_0.41                 ggthemes_5.0.0           
 [31] R6_2.5.1                  ggbeeswarm_0.7.2         
 [33] graphlayouts_1.0.2        locfit_1.5-9.8           
 [35] concaveman_1.1.0          bitops_1.0-7             
 [37] rhdf5filters_1.8.0        cachem_1.0.8             
 [39] RTriangle_1.6-0.13        DelayedArray_0.22.0      
 [41] promises_1.2.1            scales_1.3.0             
 [43] vroom_1.6.5               ggraph_2.1.0             
 [45] beeswarm_0.4.0            gtable_0.3.4             
 [47] beachmat_2.12.0           processx_3.8.3           
 [49] tidygraph_1.3.0           rlang_1.1.3              
 [51] systemfonts_1.0.5         GlobalOptions_0.1.2      
 [53] prismatic_1.1.1           yaml_2.3.8               
 [55] abind_1.4-5               httpuv_1.6.13            
 [57] tools_4.2.0               ellipsis_0.3.2           
 [59] raster_3.6-26             jquerylib_0.1.4          
 [61] proxy_0.4-27              ggridges_0.5.5           
 [63] Rcpp_1.0.12               sparseMatrixStats_1.8.0  
 [65] zlibbioc_1.42.0           classInt_0.4-10          
 [67] purrr_1.0.2               RCurl_1.98-1.14          
 [69] ps_1.7.6                  viridis_0.6.4            
 [71] cowplot_1.1.2             ggrepel_0.9.5            
 [73] fs_1.6.3                  magrittr_2.0.3           
 [75] data.table_1.14.10        magick_2.8.2             
 [77] whisker_0.4.1             hms_1.1.3                
 [79] mime_0.12                 fftwtools_0.9-11         
 [81] evaluate_0.23             xtable_1.8-4             
 [83] jpeg_0.1-10               gridExtra_2.3            
 [85] shape_1.4.6               compiler_4.2.0           
 [87] tibble_3.2.1              KernSmooth_2.23-20       
 [89] crayon_1.5.2              R.oo_1.25.0              
 [91] htmltools_0.5.7           later_1.3.2              
 [93] tzdb_0.4.0                tiff_0.1-12              
 [95] tidyr_1.3.0               DBI_1.2.1                
 [97] tweenr_2.0.2              MASS_7.3-56              
 [99] sf_1.0-15                 Matrix_1.6-5             
[101] readr_2.1.5               cli_3.6.2                
[103] R.methodsS3_1.8.2         parallel_4.2.0           
[105] cytomapper_1.9.1          igraph_1.6.0             
[107] pkgconfig_2.0.3           getPass_0.2-4            
[109] sp_2.1-2                  terra_1.7-65             
[111] scuttle_1.6.3             svglite_2.1.3            
[113] vipor_0.4.7               bslib_0.6.1              
[115] dqrng_0.3.2               XVector_0.36.0           
[117] callr_3.7.3               digest_0.6.34            
[119] rmarkdown_2.25            edgeR_3.38.4             
[121] DelayedMatrixStats_1.18.2 shiny_1.8.0              
[123] EBImage_4.38.0            rjson_0.2.21             
[125] lifecycle_1.0.4           jsonlite_1.8.8           
[127] Rhdf5lib_1.18.2           BiocNeighbors_1.14.0     
[129] viridisLite_0.4.2         limma_3.52.4             
[131] fansi_1.0.6               pillar_1.9.0             
[133] lattice_0.20-45           fastmap_1.1.1            
[135] httr_1.4.7                glue_1.7.0               
[137] png_0.1-8                 svgPanZoom_0.3.4         
[139] bit_4.0.5                 ggforce_0.4.1            
[141] class_7.3-20              stringi_1.8.3            
[143] sass_0.4.8                HDF5Array_1.24.2         
[145] nnls_1.5                  rematch2_2.1.2           
[147] dplyr_1.1.4               e1071_1.7-14             </code></pre>
</details>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.2.0 (2022-04-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] dittoSeq_1.8.1              ggplot2_3.4.4              
 [3] paletteer_1.5.0             circlize_0.4.15            
 [5] RColorBrewer_1.1-3          stringr_1.5.1              
 [7] imcRtools_1.3.5             SpatialExperiment_1.6.1    
 [9] SingleCellExperiment_1.20.1 SummarizedExperiment_1.26.1
[11] Biobase_2.56.0              GenomicRanges_1.48.0       
[13] GenomeInfoDb_1.32.4         IRanges_2.30.1             
[15] S4Vectors_0.34.0            BiocGenerics_0.42.0        
[17] MatrixGenerics_1.8.1        matrixStats_1.1.0          
[19] workflowr_1.7.1            

loaded via a namespace (and not attached):
  [1] utf8_1.2.4                shinydashboard_0.7.2     
  [3] R.utils_2.12.3            tidyselect_1.2.0         
  [5] htmlwidgets_1.6.4         grid_4.2.0               
  [7] BiocParallel_1.30.4       DropletUtils_1.16.0      
  [9] munsell_0.5.0             codetools_0.2-18         
 [11] units_0.8-5               DT_0.31                  
 [13] withr_3.0.0               colorspace_2.1-0         
 [15] highr_0.10                knitr_1.45               
 [17] rstudioapi_0.15.0         labeling_0.4.3           
 [19] git2r_0.33.0              GenomeInfoDbData_1.2.8   
 [21] polyclip_1.10-6           bit64_4.0.5              
 [23] farver_2.1.1              pheatmap_1.0.12          
 [25] rhdf5_2.40.0              rprojroot_2.0.4          
 [27] vctrs_0.6.5               generics_0.1.3           
 [29] xfun_0.41                 ggthemes_5.0.0           
 [31] R6_2.5.1                  ggbeeswarm_0.7.2         
 [33] graphlayouts_1.0.2        locfit_1.5-9.8           
 [35] concaveman_1.1.0          bitops_1.0-7             
 [37] rhdf5filters_1.8.0        cachem_1.0.8             
 [39] RTriangle_1.6-0.13        DelayedArray_0.22.0      
 [41] promises_1.2.1            scales_1.3.0             
 [43] vroom_1.6.5               ggraph_2.1.0             
 [45] beeswarm_0.4.0            gtable_0.3.4             
 [47] beachmat_2.12.0           processx_3.8.3           
 [49] tidygraph_1.3.0           rlang_1.1.3              
 [51] systemfonts_1.0.5         GlobalOptions_0.1.2      
 [53] prismatic_1.1.1           yaml_2.3.8               
 [55] abind_1.4-5               httpuv_1.6.13            
 [57] tools_4.2.0               ellipsis_0.3.2           
 [59] raster_3.6-26             jquerylib_0.1.4          
 [61] proxy_0.4-27              ggridges_0.5.5           
 [63] Rcpp_1.0.12               sparseMatrixStats_1.8.0  
 [65] zlibbioc_1.42.0           classInt_0.4-10          
 [67] purrr_1.0.2               RCurl_1.98-1.14          
 [69] ps_1.7.6                  viridis_0.6.4            
 [71] cowplot_1.1.2             ggrepel_0.9.5            
 [73] fs_1.6.3                  magrittr_2.0.3           
 [75] data.table_1.14.10        magick_2.8.2             
 [77] whisker_0.4.1             hms_1.1.3                
 [79] mime_0.12                 fftwtools_0.9-11         
 [81] evaluate_0.23             xtable_1.8-4             
 [83] jpeg_0.1-10               gridExtra_2.3            
 [85] shape_1.4.6               compiler_4.2.0           
 [87] tibble_3.2.1              KernSmooth_2.23-20       
 [89] crayon_1.5.2              R.oo_1.25.0              
 [91] htmltools_0.5.7           later_1.3.2              
 [93] tzdb_0.4.0                tiff_0.1-12              
 [95] tidyr_1.3.0               DBI_1.2.1                
 [97] tweenr_2.0.2              MASS_7.3-56              
 [99] sf_1.0-15                 Matrix_1.6-5             
[101] readr_2.1.5               cli_3.6.2                
[103] R.methodsS3_1.8.2         parallel_4.2.0           
[105] cytomapper_1.9.1          igraph_1.6.0             
[107] pkgconfig_2.0.3           getPass_0.2-4            
[109] sp_2.1-2                  terra_1.7-65             
[111] scuttle_1.6.3             svglite_2.1.3            
[113] vipor_0.4.7               bslib_0.6.1              
[115] dqrng_0.3.2               XVector_0.36.0           
[117] callr_3.7.3               digest_0.6.34            
[119] rmarkdown_2.25            edgeR_3.38.4             
[121] DelayedMatrixStats_1.18.2 shiny_1.8.0              
[123] EBImage_4.38.0            rjson_0.2.21             
[125] lifecycle_1.0.4           jsonlite_1.8.8           
[127] Rhdf5lib_1.18.2           BiocNeighbors_1.14.0     
[129] viridisLite_0.4.2         limma_3.52.4             
[131] fansi_1.0.6               pillar_1.9.0             
[133] lattice_0.20-45           fastmap_1.1.1            
[135] httr_1.4.7                glue_1.7.0               
[137] png_0.1-8                 svgPanZoom_0.3.4         
[139] bit_4.0.5                 ggforce_0.4.1            
[141] class_7.3-20              stringi_1.8.3            
[143] sass_0.4.8                HDF5Array_1.24.2         
[145] nnls_1.5                  rematch2_2.1.2           
[147] dplyr_1.1.4               e1071_1.7-14             </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
