<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="LazDaria" />

<meta name="date" content="2024-03-27" />

<title>Quality control - Part 1</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">MapMetSC</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Quality control - Part 1</h1>
<h4 class="author">LazDaria</h4>
<h4 class="date">2024-03-27</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-03-27
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>MapMetSC/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240314code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240314)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240314code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240314)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomTaschnerMandlGroupMapMettreeb1b42ebeb24eb9cb0b628f746f029e3e474bac7ctargetblankb1b42eba">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/TaschnerMandlGroup/MapMet/tree/b1b42ebeb24eb9cb0b628f746f029e3e474bac7c" target="_blank">b1b42eb</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomTaschnerMandlGroupMapMettreeb1b42ebeb24eb9cb0b628f746f029e3e474bac7ctargetblankb1b42eba"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/TaschnerMandlGroup/MapMet/tree/b1b42ebeb24eb9cb0b628f746f029e3e474bac7c" target="_blank">b1b42eb</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    analysis/figure/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/02_QC_1.Rmd</code>) and HTML
(<code>docs/02_QC_1.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/TaschnerMandlGroup/MapMet/blob/b1b42ebeb24eb9cb0b628f746f029e3e474bac7c/analysis/02_QC_1.Rmd" target="_blank">b1b42eb</a>
</td>
<td>
LazDaria
</td>
<td>
2024-03-27
</td>
<td>
02_QC_1.Rmd
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/TaschnerMandlGroup/MapMet/blob/89a61986b87ba310fb2edffcdfed9a844309a2ac/analysis/02_QC_1.Rmd" target="_blank">89a6198</a>
</td>
<td>
LazDaria
</td>
<td>
2024-03-27
</td>
<td>
03_QC_2.Rmd
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/TaschnerMandlGroup/MapMet/89a61986b87ba310fb2edffcdfed9a844309a2ac/docs/02_QC_1.html" target="_blank">89a6198</a>
</td>
<td>
LazDaria
</td>
<td>
2024-03-27
</td>
<td>
03_QC_2.Rmd
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/TaschnerMandlGroup/MapMet/blob/1d25af4dd1578bc679230ee6f515532d4dd3672e/analysis/02_QC_1.Rmd" target="_blank">1d25af4</a>
</td>
<td>
LazDaria
</td>
<td>
2024-03-25
</td>
<td>
worfklowR part1
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/TaschnerMandlGroup/MapMet/1d25af4dd1578bc679230ee6f515532d4dd3672e/docs/02_QC_1.html" target="_blank">1d25af4</a>
</td>
<td>
LazDaria
</td>
<td>
2024-03-25
</td>
<td>
worfklowR part1
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="qc-1" class="section level1">
<h1>Quality control - Exclusion of lost cells</h1>
<p>Segmentation in <a
href="https://github.com/TaschnerMandlGroup/MapMetIP">MapMetIP</a> is
performed based on high-resolution IF images. However, due to the
washing step between IF microscopy and IMC staining, some cells might be
lost. These have to be excluded before downstream analysis.</p>
<div id="load-libraries" class="section level2">
<h2>Load libraries</h2>
<details>
<summary>
Load libraries
</summary>
<pre class="r"><code>library(ggplot2)
library(bruceR)</code></pre>
<pre><code>
bruceR (v2023.9)
Broadly Useful Convenient and Efficient R functions

Packages also loaded:
✔ data.table    ✔ emmeans
✔ dplyr         ✔ lmerTest
✔ tidyr         ✔ effectsize
✔ stringr       ✔ performance
✔ ggplot2       ✔ interactions

Main functions of `bruceR`:
cc()            Describe()  TTEST()
add()           Freq()      MANOVA()
.mean()         Corr()      EMMEANS()
set.wd()        Alpha()     PROCESS()
import()        EFA()       model_summary()
print_table()   CFA()       lavaan_summary()

For full functionality, please install all dependencies:
install.packages(&quot;bruceR&quot;, dep=TRUE)

Online documentation:
https://psychbruce.github.io/bruceR

To use this package in publications, please cite:
Bao, H.-W.-S. (2023). bruceR: Broadly useful convenient and efficient R functions (Version 2023.9) [Computer software]. https://CRAN.R-project.org/package=bruceR</code></pre>
<pre class="r"><code>library(ComplexHeatmap)</code></pre>
<pre><code>Loading required package: grid</code></pre>
<pre><code>========================================
ComplexHeatmap version 2.12.1
Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/
Github page: https://github.com/jokergoo/ComplexHeatmap
Documentation: http://jokergoo.github.io/ComplexHeatmap-reference

If you use it in published research, please cite either one:
- Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional 
    genomic data. Bioinformatics 2016.
- Gu, Z. Complex Heatmap Visualization. iMeta 2022.


The new InteractiveComplexHeatmap package can directly export static 
complex heatmaps into an interactive Shiny app with zero effort. Have a try!

This message can be suppressed by:
  suppressPackageStartupMessages(library(ComplexHeatmap))
========================================</code></pre>
<pre class="r"><code>library(viridis)</code></pre>
<pre><code>Loading required package: viridisLite</code></pre>
<pre class="r"><code>library(circlize)</code></pre>
<pre><code>========================================
circlize version 0.4.15
CRAN page: https://cran.r-project.org/package=circlize
Github page: https://github.com/jokergoo/circlize
Documentation: https://jokergoo.github.io/circlize_book/book/

If you use it in published research, please cite:
Gu, Z. circlize implements and enhances circular visualization
  in R. Bioinformatics 2014.

This message can be suppressed by:
  suppressPackageStartupMessages(library(circlize))
========================================</code></pre>
<pre class="r"><code>library(RColorBrewer)
library(paletteer)
library(Rphenoannoy)</code></pre>
<pre><code>Loading required package: igraph</code></pre>
<pre><code>
Attaching package: &#39;igraph&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:circlize&#39;:

    degree</code></pre>
<pre><code>The following object is masked from &#39;package:tidyr&#39;:

    crossing</code></pre>
<pre><code>The following objects are masked from &#39;package:dplyr&#39;:

    as_data_frame, groups, union</code></pre>
<pre><code>The following objects are masked from &#39;package:stats&#39;:

    decompose, spectrum</code></pre>
<pre><code>The following object is masked from &#39;package:base&#39;:

    union</code></pre>
<pre class="r"><code>library(scuttle)</code></pre>
<pre><code>Loading required package: SingleCellExperiment</code></pre>
<pre><code>Loading required package: SummarizedExperiment</code></pre>
<pre><code>Loading required package: MatrixGenerics</code></pre>
<pre><code>Loading required package: matrixStats</code></pre>
<pre><code>
Attaching package: &#39;matrixStats&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:dplyr&#39;:

    count</code></pre>
<pre><code>
Attaching package: &#39;MatrixGenerics&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:matrixStats&#39;:

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars</code></pre>
<pre><code>Loading required package: GenomicRanges</code></pre>
<pre><code>Loading required package: stats4</code></pre>
<pre><code>Loading required package: BiocGenerics</code></pre>
<pre><code>
Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:igraph&#39;:

    normalize, path, union</code></pre>
<pre><code>The following objects are masked from &#39;package:dplyr&#39;:

    combine, intersect, setdiff, union</code></pre>
<pre><code>The following objects are masked from &#39;package:stats&#39;:

    IQR, mad, sd, var, xtabs</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    anyDuplicated, append, as.data.frame, basename, cbind, colnames,
    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
    union, unique, unsplit, which.max, which.min</code></pre>
<pre><code>Loading required package: S4Vectors</code></pre>
<pre><code>
Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:Matrix&#39;:

    expand, unname</code></pre>
<pre><code>The following objects are masked from &#39;package:data.table&#39;:

    first, second</code></pre>
<pre><code>The following object is masked from &#39;package:tidyr&#39;:

    expand</code></pre>
<pre><code>The following objects are masked from &#39;package:dplyr&#39;:

    first, rename</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    expand.grid, I, unname</code></pre>
<pre><code>Loading required package: IRanges</code></pre>
<pre><code>
Attaching package: &#39;IRanges&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:data.table&#39;:

    shift</code></pre>
<pre><code>The following objects are masked from &#39;package:dplyr&#39;:

    collapse, desc, slice</code></pre>
<pre><code>Loading required package: GenomeInfoDb</code></pre>
<pre><code>Loading required package: Biobase</code></pre>
<pre><code>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    &#39;browseVignettes()&#39;. To cite Bioconductor, see
    &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<pre><code>
Attaching package: &#39;Biobase&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:MatrixGenerics&#39;:

    rowMedians</code></pre>
<pre><code>The following objects are masked from &#39;package:matrixStats&#39;:

    anyMissing, rowMedians</code></pre>
<pre class="r"><code>library(cytomapper)</code></pre>
<pre><code>Loading required package: EBImage</code></pre>
<pre><code>
Attaching package: &#39;EBImage&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:SummarizedExperiment&#39;:

    resize</code></pre>
<pre><code>The following object is masked from &#39;package:Biobase&#39;:

    channel</code></pre>
<pre><code>The following objects are masked from &#39;package:GenomicRanges&#39;:

    resize, tile</code></pre>
<pre><code>The following objects are masked from &#39;package:IRanges&#39;:

    resize, tile</code></pre>
<pre><code>The following object is masked from &#39;package:performance&#39;:

    display</code></pre>
<pre><code>The following object is masked from &#39;package:effectsize&#39;:

    display</code></pre>
<pre><code>The following object is masked from &#39;package:data.table&#39;:

    transpose</code></pre>
<pre><code>
Attaching package: &#39;cytomapper&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:Biobase&#39;:

    channelNames, channelNames&lt;-</code></pre>
<pre class="r"><code>library(cowplot)
library(gridGraphics)</code></pre>
</details>
</div>
<div id="read-spe-object" class="section level2">
<h2>Read SPE object</h2>
<p>We will first read the spatial experiment object (SPE) generated in
the <a href="#read-data">read data section</a>. Please make sure,
<code>path</code> points to the folder where you stored the SPE
object.</p>
<pre class="r"><code>path &lt;- params$output
spe &lt;- readRDS(file.path(path,&quot;spe.rds&quot;))</code></pre>
</div>
<div id="exclude-lost-cells" class="section level2">
<h2>Exclude lost cells</h2>
<div id="quality-control-step-1" class="section level3">
<h3>Quality control step 1</h3>
<p>In a first step, we exclude cells, that have a mean intensity of zero
for all IMC channels. For that, we will use the intensities measured
based on the non-dilated masks (stored in <code>counts</code>) to avoid
capturing signals from neighboring cells. To correctly identify lost
cells in the clustering step below, we use the mean of the 20% highest
pixel intensities as a measure for marker intensity here, as this
feature is less dependent on cell size.</p>
<pre class="r"><code># Use measurements based on non-dilated mask stored in counts
intensity_mat &lt;- assay(spe, &quot;counts&quot;)
colnames(intensity_mat) &lt;- spe$sample_id

#use mean-80 intensity to avoid losing big cells
intensity_mat_mean_80 &lt;- intensity_mat[grepl(&quot;mean-80&quot;, rownames(intensity_mat)),]

#exclusion of all-zero cells
intensity_mat_imc &lt;- intensity_mat_mean_80[!grepl(&quot;IF2_GD2|IF3_CD45/56|DAPI&quot;, rownames(intensity_mat_mean_80)),]
non_zeros &lt;- as.logical(colSums(intensity_mat_imc)!=0)

spe &lt;- spe[, non_zeros]
mat &lt;- intensity_mat_mean_80[, non_zeros]</code></pre>
<p>We excluded 6121 cells in this first quality control step.</p>
<p>For the next quality control step, we will only consider IMC and
nuclear IF channels.</p>
<pre class="r"><code>mat &lt;- mat[!grepl(&quot;IF2_GD2|IF3_CD45/56&quot;, rownames(mat)),]</code></pre>
</div>
<div id="quality-control-step-2" class="section level3">
<h3>Quality control step 2</h3>
<p>Although, we excluded all cells with zero mean intensities in all IMC
channels, there might still be some cells, which actually swam off
between IF and IMC and have IMC counts due to background noise or
lateral spillover. In the heatmap below, we can clearly observe a
cluster of cells expressing the IF marker DAPI, but no IMC markers.</p>
<pre class="r"><code>#named color vectors
qual_col_pals &lt;- brewer.pal.info[brewer.pal.info$category == &#39;qual&#39;,]
col_vector_433 &lt;- colors()[grep(&#39;gr(a|e)y&#39;, grDevices::colors(), invert = T)]

sample_names &lt;- sub(&quot;_[^_]+$&quot;, &quot;&quot;, colnames(mat))
col_sample &lt;- setNames(sample(col_vector_433, length(unique(sample_names))), unique(sample_names))

set.seed(221228)
rand_cells &lt;- sample(seq_len(ncol(mat)), 2000)

column_ha &lt;- HeatmapAnnotation(sample = sample_names[rand_cells], 
                              col=list(sample=col_sample), 
                              show_legend=F)

Heatmap(mat[,rand_cells], 
        column_title = &quot;Single-cell heatmap&quot;,
        top_annotation = column_ha, 
        show_column_dend = F,
        show_column_names=F,
        col=viridis(100)
        )</code></pre>
<p><img src="figure/02_QC_1.Rmd/data-visualization-1.png" width="1056" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-data-visualization-1">
Past versions of data-visualization-1.png
</button>
</p>
<div id="fig-data-visualization-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/TaschnerMandlGroup/MapMet/blob/89a61986b87ba310fb2edffcdfed9a844309a2ac/docs/figure/02_QC_1.Rmd/data-visualization-1.png" target="_blank">89a6198</a>
</td>
<td>
LazDaria
</td>
<td>
2024-03-27
</td>
</tr>
<tr>
<td>
<a href="https://github.com/TaschnerMandlGroup/MapMet/blob/1d25af4dd1578bc679230ee6f515532d4dd3672e/docs/figure/02_QC_1.Rmd/data-visualization-1.png" target="_blank">1d25af4</a>
</td>
<td>
LazDaria
</td>
<td>
2024-03-25
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We will cluster the data to identify cells that swam off.</p>
<pre class="r"><code>set.seed(230619)

mat &lt;- t(mat)
Rphenograph_out &lt;- Rphenoannoy(mat, k = 45)</code></pre>
<pre><code>Run Rphenograph starts:
  -Input data of 2033550 rows and 42 columns
  -k is set to 45</code></pre>
<pre><code>  Finding nearest neighbors...DONE ~ 881.98 s
  Compute jaccard coefficient between nearest-neighbor sets...</code></pre>
<pre><code>Presorting knn...</code></pre>
<pre><code>presorting DONE ~ 151.725 s
  Start jaccard
DONE ~ 8.247 s
  Build undirected graph from the weighted links...DONE ~ 67.224 s
  Run louvain clustering on the graph ...DONE ~ 2115.043 s</code></pre>
<pre><code>Run Rphenograph DONE, totally takes 3072.494s.</code></pre>
<pre><code>  Return a community class
  -Modularity value: 0.8756466 
  -Number of clusters: 46</code></pre>
<pre class="r"><code>clusters &lt;- factor(membership(Rphenograph_out[[2]]))

spe$pg_clusters_lostcells &lt;- clusters</code></pre>
<p>Next, we will have a look at the mean marker expression per cluster
in a heatmap.</p>
<pre class="r"><code>mean &lt;- aggregate(mat, list(clusters), mean)
mean_t &lt;- t(mean)
mean_t &lt;- t(mean[,2:ncol(mean)])
colnames(mean_t) &lt;- mean[,&quot;Group.1&quot;]

n &lt;- length(unique(clusters))
col_clusters &lt;- setNames(sample(col_vector_433, n), unique(clusters))

column_ha = HeatmapAnnotation(cluster = mean[,&quot;Group.1&quot;], 
                              ncells=anno_barplot(as.data.frame(table(clusters))[,&quot;Freq&quot;], height = unit(3, &quot;cm&quot;)),
                              col=list(cluster=col_clusters), 
                              show_legend=T)

Heatmap(mean_t, 
        column_title = &quot;Cluster-means heatmap&quot;,
        top_annotation = column_ha, 
        cluster_columns=T,
        show_column_dend = F,
        show_column_names=T,
        show_row_names=T,
        col=viridis(100)
        )</code></pre>
<p><img src="figure/02_QC_1.Rmd/data-visualization-2-1.png" width="1056" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-data-visualization-2-1">
Past versions of data-visualization-2-1.png
</button>
</p>
<div id="fig-data-visualization-2-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/TaschnerMandlGroup/MapMet/blob/89a61986b87ba310fb2edffcdfed9a844309a2ac/docs/figure/02_QC_1.Rmd/data-visualization-2-1.png" target="_blank">89a6198</a>
</td>
<td>
LazDaria
</td>
<td>
2024-03-27
</td>
</tr>
<tr>
<td>
<a href="https://github.com/TaschnerMandlGroup/MapMet/blob/1d25af4dd1578bc679230ee6f515532d4dd3672e/docs/figure/02_QC_1.Rmd/data-visualization-2-1.png" target="_blank">1d25af4</a>
</td>
<td>
LazDaria
</td>
<td>
2024-03-25
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We can see that cluster 11 represent cells that swam off (DAPI
expression without IMC marker expression). To be sure that we only
exclude cells that are really not present on the IMC channels, we will
identify the samples with the highest number of cells from clsuter
11.</p>
<pre class="r"><code>c &lt;- &quot;11&quot;
mat &lt;- t(mat)

data &lt;- data.frame(sample=colnames(mat), clusters)
data$freq &lt;- 1

data_wide &lt;- data %&gt;% pivot_wider(names_from = c(&quot;clusters&quot;), values_from = &quot;freq&quot;, values_fn = sum, values_fill = 0)
data_long &lt;- data_wide %&gt;% pivot_longer(cols = !sample, names_to = &quot;cluster&quot;, values_to = &quot;count&quot;) 

clust &lt;- data_long[data_long$cluster==c,]
cells_per_sample &lt;- table(spe$sample_id)[clust$sample]
lost_cells_perc &lt;- clust$count/cells_per_sample
clust$percentage &lt;- lost_cells_perc
clust &lt;- clust[order(clust$percentage, decreasing=T), ]
max_sample &lt;- clust$sample[1]

write.csv(clust, file.path(params$output,&quot;lost_cells.csv&quot;))</code></pre>
<p>Sample 20220607_20220201_15-042_BM_003 has the highest number of
cells from cluster 11. For a last sanity check, we will color the lost
cells in the segmentation mask of that sample and save the colored mask
for a more thorough check in napari.</p>
<pre class="r"><code>masks &lt;- loadImages(file.path(params$input, &quot;masks&quot;, paste0(max_sample, &quot;.tif&quot;)), as.is = TRUE)
img &lt;- loadImages(file.path(params$input, &quot;img&quot;, paste0(max_sample, &quot;.tif&quot;)))

mcols(masks) &lt;- DataFrame(sample_id = names(masks))
mcols(img) &lt;- DataFrame(sample_id = names(img))
channelNames(img) &lt;- rownames(spe)[1:44]

c = 11
tmp &lt;- spe[,spe$sample_id == max_sample]
tmp$lost_cell &lt;- &quot;no&quot;
tmp[,tmp$pg_clusters_lostcells==c]$lost_cell &lt;- &quot;yes&quot;

dapi_plot &lt;- plotPixels(img, 
                       colour_by = &quot;IF1_DAPI_mean&quot;,
                       bcg = list(IF1_DAPI_mean = c(0, 1, 1)),
                       legend=NULL,
                       return_plot = TRUE)</code></pre>
<pre class="r"><code>iridium_plot &lt;- plotPixels(img, 
                       colour_by = &quot;DNA1_Ir193_mean&quot;,
                       bcg = list(DNA1_Ir193_mean = c(0, 1, 1)),
                       legend=NULL,
                       return_plot = TRUE)</code></pre>
<pre class="r"><code>mask_plot &lt;- plotCells(masks,
                       object = tmp, 
                       cell_id = &quot;ObjectNumber&quot;, 
                       img_id = &quot;sample_id&quot;,
                       colour_by = &quot;lost_cell&quot;,
                       colour = list(lost_cell = c(&quot;no&quot; = &quot;gray&quot;, &quot;yes&quot; = &quot;red&quot;)),
                       missing_colour = &quot;white&quot;,
                       legend=NULL,
                       return_plot = TRUE)</code></pre>
<pre class="r"><code>p1 &lt;- ggdraw(dapi_plot$plot, clip = &quot;on&quot;)
p2 &lt;- ggdraw(iridium_plot$plot, clip = &quot;on&quot;)
p3 &lt;- ggdraw(mask_plot$plot, clip = &quot;on&quot;)

plot_grid(p1, p2, p3)</code></pre>
<p><img src="figure/02_QC_1.Rmd/plot-grid-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>By comparing the DAPI image, the Iridium image and the binary mask
(highlighting cells from cluster 11 in blue), we see that cluster 11
definitely represents the cluster of lost cells. We excluded 30285 cells
in total or 46.6640986 cells per image (n=649).</p>
</div>
</div>
<div id="save-object" class="section level2">
<h2>Save object</h2>
<p>Finally, we exclude the lost cells from the SPE object and save the
latter.</p>
<pre class="r"><code>spe &lt;- spe[, spe$pg_clusters_lostcells!=11]
saveRDS(spe, file.path(params$output, &quot;spe_qc1.rds&quot;))</code></pre>
</div>
<div id="session-info" class="section level2">
<h2>Session Info</h2>
<details>
<summary>
SessionInfo
</summary>
<pre><code>R version 4.2.0 (2022-04-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    grid      stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] gridGraphics_0.5-1          cowplot_1.1.2              
 [3] cytomapper_1.9.1            EBImage_4.38.0             
 [5] scuttle_1.6.3               SingleCellExperiment_1.20.1
 [7] SummarizedExperiment_1.26.1 Biobase_2.56.0             
 [9] GenomicRanges_1.48.0        GenomeInfoDb_1.32.4        
[11] IRanges_2.30.1              S4Vectors_0.34.0           
[13] BiocGenerics_0.42.0         MatrixGenerics_1.8.1       
[15] matrixStats_1.1.0           Rphenoannoy_0.1.0          
[17] igraph_1.6.0                paletteer_1.5.0            
[19] RColorBrewer_1.1-3          circlize_0.4.15            
[21] viridis_0.6.4               viridisLite_0.4.2          
[23] ComplexHeatmap_2.12.1       interactions_1.1.5         
[25] lmerTest_3.1-3              lme4_1.1-35.1              
[27] Matrix_1.6-5                performance_0.10.8         
[29] effectsize_0.8.6            emmeans_1.9.0              
[31] data.table_1.14.10          stringr_1.5.1              
[33] tidyr_1.3.0                 dplyr_1.1.4                
[35] bruceR_2023.9               ggplot2_3.4.4              
[37] workflowr_1.7.1            

loaded via a namespace (and not attached):
  [1] utf8_1.2.4                shinydashboard_0.7.2     
  [3] R.utils_2.12.3            tidyselect_1.2.0         
  [5] htmlwidgets_1.6.4         BiocParallel_1.30.4      
  [7] DropletUtils_1.16.0       jtools_2.2.2             
  [9] munsell_0.5.0             codetools_0.2-18         
 [11] withr_3.0.0               colorspace_2.1-0         
 [13] highr_0.10                knitr_1.45               
 [15] rstudioapi_0.15.0         git2r_0.33.0             
 [17] GenomeInfoDbData_1.2.8    datawizard_0.9.1         
 [19] rhdf5_2.40.0              rprojroot_2.0.4          
 [21] coda_0.19-4               vctrs_0.6.5              
 [23] generics_0.1.3            TH.data_1.1-2            
 [25] xfun_0.41                 R6_2.5.1                 
 [27] doParallel_1.0.17         ggbeeswarm_0.7.2         
 [29] clue_0.3-65               locfit_1.5-9.8           
 [31] bitops_1.0-7              rhdf5filters_1.8.0       
 [33] cachem_1.0.8              DelayedArray_0.22.0      
 [35] promises_1.2.1            scales_1.3.0             
 [37] multcomp_1.4-25           beeswarm_0.4.0           
 [39] gtable_0.3.4              beachmat_2.12.0          
 [41] Cairo_1.6-2               processx_3.8.3           
 [43] sandwich_3.1-0            rlang_1.1.3              
 [45] systemfonts_1.0.5         GlobalOptions_0.1.2      
 [47] splines_4.2.0             yaml_2.3.8               
 [49] abind_1.4-5               httpuv_1.6.13            
 [51] tools_4.2.0               SpatialExperiment_1.6.1  
 [53] ellipsis_0.3.2            raster_3.6-26            
 [55] jquerylib_0.1.4           Rcpp_1.0.12              
 [57] sparseMatrixStats_1.8.0   zlibbioc_1.42.0          
 [59] purrr_1.0.2               RCurl_1.98-1.14          
 [61] ps_1.7.6                  GetoptLong_1.0.5         
 [63] zoo_1.8-12                cluster_2.1.3            
 [65] fs_1.6.3                  magrittr_2.0.3           
 [67] magick_2.8.2              RANN_2.6.1               
 [69] mvtnorm_1.2-4             whisker_0.4.1            
 [71] mime_0.12                 evaluate_0.23            
 [73] fftwtools_0.9-11          xtable_1.8-4             
 [75] jpeg_0.1-10               gridExtra_2.3            
 [77] shape_1.4.6               compiler_4.2.0           
 [79] tibble_3.2.1              crayon_1.5.2             
 [81] minqa_1.2.6               R.oo_1.25.0              
 [83] htmltools_0.5.7           later_1.3.2              
 [85] tiff_0.1-12               DBI_1.2.1                
 [87] MASS_7.3-56               boot_1.3-28              
 [89] cli_3.6.2                 R.methodsS3_1.8.2        
 [91] parallel_4.2.0            insight_0.19.7           
 [93] pkgconfig_2.0.3           getPass_0.2-4            
 [95] numDeriv_2016.8-1.1       sp_2.1-2                 
 [97] terra_1.7-65              foreach_1.5.2            
 [99] svglite_2.1.3             vipor_0.4.7              
[101] bslib_0.6.1               dqrng_0.3.2              
[103] XVector_0.36.0            estimability_1.4.1       
[105] callr_3.7.3               digest_0.6.34            
[107] parameters_0.21.3         rmarkdown_2.25           
[109] edgeR_3.38.4              DelayedMatrixStats_1.18.2
[111] shiny_1.8.0               rjson_0.2.21             
[113] nloptr_2.0.3              lifecycle_1.0.4          
[115] nlme_3.1-157              jsonlite_1.8.8           
[117] Rhdf5lib_1.18.2           limma_3.52.4             
[119] fansi_1.0.6               pillar_1.9.0             
[121] lattice_0.20-45           fastmap_1.1.1            
[123] httr_1.4.7                survival_3.3-1           
[125] glue_1.7.0                bayestestR_0.13.1        
[127] png_0.1-8                 iterators_1.0.14         
[129] svgPanZoom_0.3.4          pander_0.6.5             
[131] stringi_1.8.3             sass_0.4.8               
[133] HDF5Array_1.24.2          nnls_1.5                 
[135] rematch2_2.1.2           </code></pre>
</details>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.2.0 (2022-04-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    grid      stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] gridGraphics_0.5-1          cowplot_1.1.2              
 [3] cytomapper_1.9.1            EBImage_4.38.0             
 [5] scuttle_1.6.3               SingleCellExperiment_1.20.1
 [7] SummarizedExperiment_1.26.1 Biobase_2.56.0             
 [9] GenomicRanges_1.48.0        GenomeInfoDb_1.32.4        
[11] IRanges_2.30.1              S4Vectors_0.34.0           
[13] BiocGenerics_0.42.0         MatrixGenerics_1.8.1       
[15] matrixStats_1.1.0           Rphenoannoy_0.1.0          
[17] igraph_1.6.0                paletteer_1.5.0            
[19] RColorBrewer_1.1-3          circlize_0.4.15            
[21] viridis_0.6.4               viridisLite_0.4.2          
[23] ComplexHeatmap_2.12.1       interactions_1.1.5         
[25] lmerTest_3.1-3              lme4_1.1-35.1              
[27] Matrix_1.6-5                performance_0.10.8         
[29] effectsize_0.8.6            emmeans_1.9.0              
[31] data.table_1.14.10          stringr_1.5.1              
[33] tidyr_1.3.0                 dplyr_1.1.4                
[35] bruceR_2023.9               ggplot2_3.4.4              
[37] workflowr_1.7.1            

loaded via a namespace (and not attached):
  [1] utf8_1.2.4                shinydashboard_0.7.2     
  [3] R.utils_2.12.3            tidyselect_1.2.0         
  [5] htmlwidgets_1.6.4         BiocParallel_1.30.4      
  [7] DropletUtils_1.16.0       jtools_2.2.2             
  [9] munsell_0.5.0             codetools_0.2-18         
 [11] withr_3.0.0               colorspace_2.1-0         
 [13] highr_0.10                knitr_1.45               
 [15] rstudioapi_0.15.0         git2r_0.33.0             
 [17] GenomeInfoDbData_1.2.8    datawizard_0.9.1         
 [19] rhdf5_2.40.0              rprojroot_2.0.4          
 [21] coda_0.19-4               vctrs_0.6.5              
 [23] generics_0.1.3            TH.data_1.1-2            
 [25] xfun_0.41                 R6_2.5.1                 
 [27] doParallel_1.0.17         ggbeeswarm_0.7.2         
 [29] clue_0.3-65               locfit_1.5-9.8           
 [31] bitops_1.0-7              rhdf5filters_1.8.0       
 [33] cachem_1.0.8              DelayedArray_0.22.0      
 [35] promises_1.2.1            scales_1.3.0             
 [37] multcomp_1.4-25           beeswarm_0.4.0           
 [39] gtable_0.3.4              beachmat_2.12.0          
 [41] Cairo_1.6-2               processx_3.8.3           
 [43] sandwich_3.1-0            rlang_1.1.3              
 [45] systemfonts_1.0.5         GlobalOptions_0.1.2      
 [47] splines_4.2.0             yaml_2.3.8               
 [49] abind_1.4-5               httpuv_1.6.13            
 [51] tools_4.2.0               SpatialExperiment_1.6.1  
 [53] ellipsis_0.3.2            raster_3.6-26            
 [55] jquerylib_0.1.4           Rcpp_1.0.12              
 [57] sparseMatrixStats_1.8.0   zlibbioc_1.42.0          
 [59] purrr_1.0.2               RCurl_1.98-1.14          
 [61] ps_1.7.6                  GetoptLong_1.0.5         
 [63] zoo_1.8-12                cluster_2.1.3            
 [65] fs_1.6.3                  magrittr_2.0.3           
 [67] magick_2.8.2              RANN_2.6.1               
 [69] mvtnorm_1.2-4             whisker_0.4.1            
 [71] mime_0.12                 evaluate_0.23            
 [73] fftwtools_0.9-11          xtable_1.8-4             
 [75] jpeg_0.1-10               gridExtra_2.3            
 [77] shape_1.4.6               compiler_4.2.0           
 [79] tibble_3.2.1              crayon_1.5.2             
 [81] minqa_1.2.6               R.oo_1.25.0              
 [83] htmltools_0.5.7           later_1.3.2              
 [85] tiff_0.1-12               DBI_1.2.1                
 [87] MASS_7.3-56               boot_1.3-28              
 [89] cli_3.6.2                 R.methodsS3_1.8.2        
 [91] parallel_4.2.0            insight_0.19.7           
 [93] pkgconfig_2.0.3           getPass_0.2-4            
 [95] numDeriv_2016.8-1.1       sp_2.1-2                 
 [97] terra_1.7-65              foreach_1.5.2            
 [99] svglite_2.1.3             vipor_0.4.7              
[101] bslib_0.6.1               dqrng_0.3.2              
[103] XVector_0.36.0            estimability_1.4.1       
[105] callr_3.7.3               digest_0.6.34            
[107] parameters_0.21.3         rmarkdown_2.25           
[109] edgeR_3.38.4              DelayedMatrixStats_1.18.2
[111] shiny_1.8.0               rjson_0.2.21             
[113] nloptr_2.0.3              lifecycle_1.0.4          
[115] nlme_3.1-157              jsonlite_1.8.8           
[117] Rhdf5lib_1.18.2           limma_3.52.4             
[119] fansi_1.0.6               pillar_1.9.0             
[121] lattice_0.20-45           fastmap_1.1.1            
[123] httr_1.4.7                survival_3.3-1           
[125] glue_1.7.0                bayestestR_0.13.1        
[127] png_0.1-8                 iterators_1.0.14         
[129] svgPanZoom_0.3.4          pander_0.6.5             
[131] stringi_1.8.3             sass_0.4.8               
[133] HDF5Array_1.24.2          nnls_1.5                 
[135] rematch2_2.1.2           </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
